# MOD 8 and Neighborly 2n+1

- The lattice at any given generation is almost evenly divided into $2n+1$ *neighborly* and *non-neighborly* partitions based on the label
- The integer partitioning $\pmod{8}$ with respect to neighborlyness are not quite so evenly distributed -- only 3/8ths of all integers are neighborly:

| MOD 8 | Partition | Peer | $2n+1$ | $\frac{n-1}{2}$ | Comment |
| --- | --- | --- | --- |--- | --- |
| 0 | non-neighborly | N/A | $1 \pmod{8}$ : neighborly| __non-integer__ | $N \equiv 0 \pmod{4}$  |
| 1 | neighborly  | $2n+1$ | $3 \pmod{16}$ : non-neighborly<br/> $11 \pmod{16}$ : neighborly<br/>| $0 \pmod{8}$ <br> $4 \pmod{8}$|  |
| 2 | neighborly  | $2n+1$ | $5 \pmod{8}$ : non-neighborly | (non-integer) | $N \equiv 2 \pmod{4}$ |
| 3 | non-neighborly | $\frac{n-1}{2}$ <br>None | $7 \pmod{8}$ : neighborly | $1 \pmod{8}$ : neighborly <br> $5 \pmod{8}  $ : non-neighborly | $N \equiv 3 \pmod{4}$ , $2(2n+1) + 1 = 4n+3$|
| 4 | non-neighborly  | N/A |  $1 \pmod{8}$ : neighborly |   __non-integer__  | $N \equiv 0 \pmod{4}$ |
| 5 | non-neighborly  | $\frac{n-1}{2}$  | $2 \pmod{8}$ : neighborly |  PEER $2 \lor 6  \pmod{8}$:  neighborly  |  |
| 6 | neighborly  | $2n+1$ | $5 \pmod{8}$ : non-neighborly|   (non-integer)   | $N \equiv 2 \pmod{4}$ |
| 7 | EITHER | Mix | __SELF__ | __SELF__  | $N \equiv 3 \pmod{4}$ , $2(2n+1) + 1 = 4n+3$|

So the lattice generates neighborly numbers faster than than their overall occurence as a simple paritioning of the integers.  This will be because the magnitude of non-neighborly numbers (e.g. $2^n$) grow faster at the top of the lattice and so wind up being spread out across more generations.

## $7 \pmod{8}$ Result is interesting

Looks like half of all $7 \pmod{8}$ are neighborly or non-neighborly, but they are not patitioned by $\pmod{16}$

## $3 \pmod{8}$ Result is unexpected

This breaks my earlier belief that all integers had a $n' -> 2n+1$ *neighborly* peer.  Half of all $3 \pmod{8}$ (even though odd) do not have an easy-to-find $\frac{(3 \pmod{8})-1}{2}$ nieghborly peer. 

The similarities and differences between $3 \pmod{8}$ and $7 \pmod{8}$ are also notable since both are related to  $3 \pmod{4}$ which is the result of applying $n' =2n+1$ twice.

## $0 \pmod{8}, 4 \pmod{8}$ Results should have been expected.  

These numbers being even and non-neighborly could never have a neighborly 2n+1 preceding peer integer.  Some (All?) of these non-integers are probably non-lattice.

In general, the 2n+1 operation does not provide the clean partitioning I thought it did where we would always find pairs of neighborly and non-neighborly numbers.

But other facts of the partitioning hold:  

-  At a given generation, the lattice is nearly evenly divided between neighborly ($2^{a-1} - 1$ nodes) and non-neighborly ($2^{a-1} + 1$ nodes) rationals/integers
-  3/8 of all integers are neighborly, neighborly numbers always have a peer non-neighborly number
-  5/8 of all integers are non-neighborly
    -  3 out of 10 of these numbers have an easily found reverse mapping to a preceding 2n+1 number
- 9/16 of all integers have a peer (e.g. are either $n$ or $n'$ in $n' = 2n+1$ and their peer is the other number) under the **neighborly $2n+1$** operation

# $7 \pmod{8} $ Numbers

## The $7 \pmod{8} $ numbers are closed under $n' = 2n+ 1$

### 1. Assumption

Let $ n \equiv 7 \pmod{8} $.  
This means:
$$
n = 8k + 7 \quad \text{for some integer } k.
$$

### 2. Apply the transformation

$$
n' = 2n + 1 = 2(8k + 7) + 1 = 16k + 14 + 1 = 16k + 15.
$$


### 3. Reduce modulo 8

$4
n' \bmod 8 = (16k + 15) \bmod 8.
$$

Since $16k$ is divisible by $8$:
$$
16k \bmod 8 = 0.
$$

So:
$$
n' \bmod 8 = 15 \bmod 8 = 7.
$$

### 4. Conclusion

$$
n' \equiv 7 \pmod{8}.
$$

Thus, the set of numbers congruent to $7 \pmod{8}$ is **closed under the operation** $n' = 2n + 1$.

## We also find that 1/2 of $7 \pmod{8} $ Numbers are Neighborly under $n' = 2n+1$

So we know that $7 \pmod{8} $ Numbers come in far-flung pairs in the lattice.



In [3]:
import sys, io
import math
import numpy as np
import pandas as pd
import random
import re
from scipy.optimize import nnls
from fractions import Fraction
from sympy import factorint
from itertools import product
from sympy.ntheory import factorint
from typing import List, Optional, Tuple

# $4n+3$ Numbers in the lattice and $\pmod{8}$

## Forward:
### $F_0$

Always gives odd numerator 1,3,5,7 (mod 8)

### $F_1$

Always gives odd numerator  1,3,5,7 (mod 8)

## Reverse:
### $\frac{3n+1}{2}$
$\frac{3n+1}{2}$ applied to a $4n+3$ number can have the following $\pmod{8}$ outcomes:

- 1: neighborly with peer
- 3: non-neighborly with peer half the time (a 4n+1 number)
- 5: non-neighborly with neighborly peer
- 7: BOTH, with a peer (a 4n+1 number)

### $\frac{n}{2}$ N/A for Odd Number

In [4]:
"""
Code from previous notebooks,

TODO: Remove unneed functions not needed for mrTup representation

"""

def T(n):
    """ Compute next value in simplified Collatz sequence.
    """
    if n & 1 == 0:
        return n//2
    else:
        return (3*n + 1)//2
#
def L_T(n):
    """ Compute binary label-string for a given Collatz number
    """
    if n == 1:
        return "1"

    S = ""
    while n != 1:
        if n & 1 == 0:
            n = n//2
            S = S + "1"
        else:
            n = (3*n + 1)//2
            S = S + "0"
    return S
#
def Ay_L(L):
    """ Generate A matrix and y vector from label string
    """
    rank = len(L) + 2    
    A = np.zeros((rank,rank))
    y = np.zeros((rank))
    for row in range(rank-3):
        if L[row] == "1":
            a_val = -1.0
            y_val = 0.0
        else:
            a_val = -3.0
            y_val = 1.0
        A[row][row] = a_val
        A[row][row+1] = 2.0
        y[row] = y_val
    #
    # Last 3 rows are always the same
    row = rank - 3
    A[row][row] = -1
    A[row][row+1] = 2
    y[row] = 0
    row = rank - 2
    A[row][row] = -3
    A[row][row+1] = 2
    y[row] = 1
    row = rank - 1
    A[row][row] = -1
    A[row][row-2] = 1
    y[row] = 0
    
    return A, y
#
def solve_Ay_L(L):
    """ Solve for the x vector given the label-string
    """
    A, y = Ay_L(L)
    return A, np.linalg.solve(A, y), y
#
def x0_L(L):
    """ Get the x[0] value given a label-string
    """
    A, x, y = solve_Ay_L(L)
    return round(x[0])  # clean up mantisa garbage
#

def countZeros(label):
    zero_count = 0
    for bit in label:
        if bit == "0":
            zero_count += 1
    return zero_count
#

def Z(L):
    """ Indexes of zeros in label string
    """
    for i in range(len(L)):
        if L[i] == "0":
            yield i
#
def a_b_c_L(L):
    """ Get the (power-of-two, power-of-three, zero-sum-accumulator) tuple for a node given its label
    """
    a = len(L)
    b = 0
    for bit in L:
        if bit == "0":
            b += 1
    ZZ = [(j,i) for j, i in enumerate(Z(L))]
    c = sum((3 ** (b - j - 1)) * (2 ** (i)) for j, i in ZZ)
    S = [zz[1] for zz in ZZ]
    return (a,b,c,S)
#
def val_a_b_c(a_b_c):
    """ Get the value for a node given the tuple (power-of-two, power-of-three, zero-sum-accumulator)
    """
    a, b, c = a_b_c
    f = Fraction( ((2**a) - c), (3**b) )
    return (f.numerator, f.denominator)
#
def val_a_b_c_L(a_b_c_L):
    a_b_c = a_b_c_L[0:3]
    return val_a_b_c(a_b_c)
#
def val_L(L):
    """ Get the value for a node given the label string
    """
    return val_a_b_c_L(a_b_c_L(L))
#
def collatzPath(collatzNumber):
    path = []
    while collatzNumber != 1:
        if (collatzNumber & 1) == 0:
            collatzNumber = collatzNumber // 2
            path.append("1")
        else:
            collatzNumber = (3 * collatzNumber + 1) // 2
            path.append("0")
    return "".join(path)
#

def collatzPath2(n_d_tup, truncate_at=100):
    """
    Collatz Path for rationals with loop detection and "too long" cut-off
    """
    chain = [n_d_tup]
    path = []
    governor = truncate_at
    while n_d_tup != (1, 1):
        if (n_d_tup[0] & 1) == 0:
            n_d_tup = (n_d_tup[0]//2, n_d_tup[1])
            path.append("1")
        else:
            f = Fraction((3 * n_d_tup[0] + n_d_tup[1])//2, n_d_tup[1])
            n_d_tup = (f.numerator, f.denominator)
            path.append("0")
        if n_d_tup in chain:
            path.insert(0,"↺")
            break
        
        governor -= 1
        if governor == 0:
            path.insert(0,"∀")
            break
        #
        chain.append(n_d_tup)
    #
    return ("".join(path), chain)
#

    
#
#  Mixed Radix form and functions
#

N_ = ((0,0), [])

def mr_TupItemValue(a_b, a_0):
    a,b = a_b
    val = (2**a)*(3**(a_0 + b))
    # print(f"val {a_b}*(3**{a_0}) = {val}")
    return val
#
def mrTupValue(mr_tup):
    # Multiplying the numerator by 3 ** the generation keeps us in integer land
    a_0 = mr_tup[0][0]
    total = mr_TupItemValue(mr_tup[0], a_0)
    for a_b in mr_tup[1]:
        total -= mr_TupItemValue(a_b, a_0)
    frac = Fraction(total, 3**a_0)
    return (frac.numerator, frac.denominator)
#

def F_0(mr_tup):
    return ( (mr_tup[0][0]+1, mr_tup[0][1]-1), mr_tup[1] + [(mr_tup[0][0], -(len(mr_tup[1])+1))] )
#

def F_1(mr_tup):
    return ((mr_tup[0][0]+1, mr_tup[0][1]), mr_tup[1])
#
def mrTupFromPath(label):
    mr_tup = N_
    for bit in label:
        if bit == "1":
            mr_tup = F_1(mr_tup)
        else:
            mr_tup = F_0(mr_tup)
    return mr_tup
#
def mrTupToPath(T):
    """
    Convert mrTup to node label
    The T[1] list encodes the positions of the zeros in the numerator power of two values
    """
    S = ["1"] * T[0][0]
    for j in range(len(T[1])):
        S[(T[1][j][0])] = "0" 
    return "".join(S)
#

def mrTupFromValue(n):
    label = collatzPath(n)
    return mrTupFromPath(label)
#

def strip_01(label):
    while len(label) > 2 and label[-2:] == "01":
        label = label[0:-2]
    return label
#

LABEL_RX = re.compile('^(?P<prefix>.*?)((?P<inttag>111)(?P<tail>((01)*)))$')

def split_int_label(s):
    """
    Splits label into 3 parts and returns prefix and suffix if matches integer-candidate pattern
    """
    match = LABEL_RX.search(s)
    if match:
        return (match.group('prefix'), match.group('tail'))
    else:
        return None  # or raise an error if preferred
#

def generationLabels(a):
    if a == 0:
        return ""
    seqs = product('10', repeat=(a))
    for bit_tup in seqs:
        label = "".join(bit_tup)
        yield label
#
def generationTups(a):
    for label in generationLabels(a):
        mrTup = mrTupFromPath(label)
        yield (label, mrTup, mrTupValue(mrTup))
    
def generationIntCandidateLabels(a):
    seqs = product('10', repeat=(a))
    for bit_tup in seqs:
        label = "".join(bit_tup)
        head_tail = split_int_label(label)
        if head_tail:
            # return int candidate with stripped tail
            yield head_tail[0] + "111"
#

def mrIntTupsForGeneration(aa):
    for label in generationIntCandidateLabels(aa):
        mrTup = mrTupFromPath(label)
        yield (label, mrTup, mrTupValue(mrTup))
#
def generationGenNums(a):
    vals = []
    bb = 3**(a)
    for infoTup in generationTups(a):
        _, __, val_tup = infoTup
        vals.append(val_tup[0] * (bb//val_tup[1]))
    vals.sort(reverse=True)
    for idx, val in enumerate(vals):
        if idx < len(vals) - 1:
            delta = val - vals[idx+1]
        else:
            delta = None
        print(f'{val}\t{delta}')
#        
def generationPairGenNums(a):
    vals = []
    bb = 3**(a+1)
    for infoTup in generationTups(a):
        _, __, val_tup = infoTup
        vals.append(val_tup[0] * (bb//val_tup[1]))
    for infoTup in generationTups(a+1):
        _, __, val_tup = infoTup
        vals.append(val_tup[0] *  (bb//val_tup[1]))
    vals.sort(reverse=True)
    for idx, val in enumerate(vals):
        if idx < len(vals) - 1:
            delta = val - vals[idx+1]
        else:
            delta = None
        print(f'{val}\t{delta}')
#
def mr2Nplus_1(T):
    B = len(T[1])  
    L = [(0, -1)]

    # Keep initial zeros
    idx = 0
    for idx, val in enumerate(T[1]):
        if T[1][idx][0] != idx:
            break
        L.append( (T[1][idx][0] + 1, T[1][idx][1]-1) )
    # Remove the first tuple where (a, -a) is true
    match = False
    for i in range(idx, B, 1):
        if (not match) and (T[1][i][0] == -T[1][i][1]):
            match = True
        else:
            L.append( (T[1][i][0]+1, T[1][i][1]) )
    if not match:
        return None
    return ( (T[0][0] + 1, T[0][1]), L)
#

'''
Do not think this works yet:
def mr2Nplus_1_inv(T):
    """
    Find the inverse tuple of mr2Nplus_1

    We know the zero was removed when L item index matched numerator and denominator
    """
    B = len(T[1])  
    L = []

    # Keep initial zeros after poping (0, -1) off the front
    idx = 0
    for idx, val in enumerate(T[1]):
        if idx == 0:
            if val == (0, -1):
                continue
            else:
                # This tuple was not generated by mr2Nplus_1
                return None
        else:
            L.append( (T[1][idx][0] - 1, T[1][idx][1] + 1) )
    inserted = False
    for i in range(idx, B, 1):
        if (not inserted) and i == (len(L)+1) and (T[0][i][0] > i):
            # Next zero in 2n+1 L list is larger than i ... insert the (i, -i) term
            L.append( (i, -i) )    
            inserted == True
        L.append( (T[1][i][0]-1, T[1][i][1]) )
        
    return ( (T[0][0] - 1, T[0][1]), L)
#
'''

def lattice2N_plus1_pairs(a):
    """
    For a given depth in the tree, generate all 2n+1 pairs in the tree
    """
    seqs = product('10', repeat=(a))
    for bit_tup in seqs:
        label = "".join(bit_tup)
        label = strip_01(label)
        val = mrTupValue(mrTupFromPath(label))
        f = Fraction(2 * val[0] + 1, val[1])
        val_ = (f.numerator, f.denominator)
        label_, chain_ = collatzPath2(val_)
        if (len(label_) == 0):
            d = len(label)
        elif label_[0] in ["↺", "∀"]:
            d = 100 + len(label)
        else:
            d = distance(label, label_)
        yield (len(label), d, (val, label), (val_, label_))
#

def mrTupToLaTex(T):
    a, b = T[0]
    s = "\\frac{2^{%d}}{3^{%d}}"%(a, -b)
    L = T[1]
    if len(L) > 0:
        s = s + " - ( "
        plus = "  "
        for c_d in L:
            c, d = c_d
            t = "\\frac{2^{%d}}{3^{%d}}"%(c, -d)
            s = s + plus + t
            plus = " + "
        s = s + " )"
    return "$ " + s + " $"
#

In [4]:
def checkTupEasy(label):
    if len(label) < 4:
        label = label + "0101"
    T = mrTupFromPath(label)
    TT = mr2Nplus_1(T)
    if TT is not None:
        return True
    else:
        return False
#

In [5]:
# No label = "11+" lattice nodes are "easy" and only 2 -> 5 is somewhat neighborly
for i in range(0, 20):
    val = 2*(2**i) + 1
    tup_2n_plus1 =  mrTupFromValue(val)    
    print ((i, val, mrTupToPath(tup_2n_plus1), tup_2n_plus1))

(0, 3, '00111', ((5, -2), [(0, -1), (1, -2)]))
(1, 5, '0111', ((4, -1), [(0, -1)]))
(2, 9, '0100010110111', ((13, -6), [(0, -1), (2, -2), (3, -3), (4, -4), (6, -5), (9, -6)]))
(3, 17, '010110111', ((9, -3), [(0, -1), (2, -2), (5, -3)]))
(4, 33, '010100110010110111', ((18, -8), [(0, -1), (2, -2), (4, -3), (5, -4), (8, -5), (9, -6), (11, -7), (14, -8)]))
(5, 65, '0101011100010110111', ((19, -8), [(0, -1), (2, -2), (4, -3), (8, -4), (9, -5), (10, -6), (12, -7), (15, -8)]))
(6, 129, '01010100011000010100100010000101100010010000001100001110101011101100011110111', ((77, -44), [(0, -1), (2, -2), (4, -3), (6, -4), (7, -5), (8, -6), (11, -7), (12, -8), (13, -9), (14, -10), (16, -11), (18, -12), (19, -13), (21, -14), (22, -15), (23, -16), (25, -17), (26, -18), (27, -19), (28, -20), (30, -21), (33, -22), (34, -23), (35, -24), (37, -25), (38, -26), (40, -27), (41, -28), (42, -29), (43, -30), (44, -31), (45, -32), (48, -33), (49, -34), (50, -35), (51, -36), (55, -37), (57, -38), (59, -39), (63, -40

In [6]:
mrTupFromPath("11111"), mrTupValue(mrTupFromPath("11111"))



(((5, 0), []), (32, 1))

In [7]:
PARTS_RX_tail = re.compile('^(?P<head>[01]*?)((?P<inttag>(111)?)(?P<tail>((01)*)))$')
PARTS_RX_head = re.compile('^(?P<prefix>0*)(?P<ones>1*)(?P<thezero>0?)(?P<remainder>[01]*)$')
RULE1_RX = re.compile('^(?P<prefix>0*)(?P<ones>1+)(?P<thezero>0)(?P<remainder>[01]*)((?P<inttag>111)(?P<tail>((01)*)))$')

In [8]:
def mrTupLabelParts(label):
    match_tail = PARTS_RX_tail.search(label)
    if match_tail:
        match_head = PARTS_RX_head.search(match_tail.group('head'))
        if match_head:
            P = (match_head.group('prefix'), match_head.group('ones'), match_head.group('thezero'), match_head.group('remainder'), match_tail.group('inttag'), match_tail.group('tail'))
    else:
        P = ("","","",label, "", "")
    return P
#

In [9]:
def genEasy2n_plus_1(a):
    """
    Generate all "easy" 2n+1 mrTups
    """
    if a <= 5:
        yield None
    else:
        for label in generationLabels(a):
            T = mrTupFromPath(label)
            val = mrTupValue(T)
            if val[1] == 1:
                did_match = False
                match = RULE1_RX.search(label)
                if match:
                    if a <= 12:
                        print((match.group('prefix'), match.group('ones'), match.group('thezero'), match.group('remainder'), match.group('inttag'), match.group('tail')))
                    did_match = True
                    T_ = mr2Nplus_1(T)
                    if T_ is None:
                        print("EASY NOT EASY:")
                        print((val[0], label))
                    else:
                        val_ = mrTupValue(T_)
                        if  val_[1] != 1 or val_[0] != (2*val[0] + 1):
                            print("EASY NOT EASY -- check fail:")
                            print((val[0], label))
                yield((label, val, did_match))
#

In [10]:
def mr2Nplus_1(T):
    B = len(T[1])  
    L = [(0, -1)]

    # Keep initial zeros
    idx = 0
    for idx, val in enumerate(T[1]):
        if T[1][idx][0] != idx:
            break
        L.append( (T[1][idx][0] + 1, T[1][idx][1]-1) )
    # Remove the first tuple where (a, -a) is true
    match = False
    for i in range(idx, B, 1):
        if (not match) and (T[1][i][0] == -T[1][i][1]):
            match = True
        else:
            L.append( (T[1][i][0]+1, T[1][i][1]) )
    if not match:
        return None
    return ( (T[0][0] + 1, T[0][1]), L)
#


In [11]:
def splitOutInitialZeros(L):
    L_a = []
    L_z = []
    initial = True
    for i in range(len(L)):
        if initial and L[i][0] == i:
            L_a.append(L[i])
        else:
            initial = False
            L_z.append(L[i])
    return L_a, L_z
#


In [12]:
T_27 = mrTupFromValue(27)
print(splitOutInitialZeros(T_27[1]))

([(0, -1), (1, -2)], [(3, -3), (4, -4), (5, -5), (6, -6), (7, -7), (9, -8), (11, -9), (12, -10), (14, -11), (15, -12), (16, -13), (18, -14), (19, -15), (20, -16), (21, -17), (23, -18), (26, -19), (27, -20), (28, -21), (30, -22), (31, -23), (33, -24), (34, -25), (35, -26), (36, -27), (37, -28), (38, -29), (41, -30), (42, -31), (43, -32), (44, -33), (48, -34), (50, -35), (52, -36), (56, -37), (59, -38), (60, -39), (61, -40), (66, -41)])


In [13]:

def mrTup2NIdentityElement(T):
    """ 
    returns the index of the L list item. if any, that allows
    a 2n operation to be transformed into a 2n+1 operation by
    the removal of a single element.

    Brute force for now
    """
    val = mrTupValue(T)
    f = Fraction(2*val[0] + val[1], val[1])
    val_goal = (f.numerator, f.denominator)

    a = T[0][0]
    b = T[0][1]
    L = T[1]

    L_head, L_tail = splitOutInitialZeros(L)

    L_head_ = [(0, -1)]
    for c_d in L_head:
        L_head_.append((c_d[0] + 1, c_d[1] -1))
    
    for i in range(len(L_tail)):
        # which one of these, when we ditch it gives us our goal value?
        L_tail_ = [(c_d[0] + 1, c_d[1]) for idx, c_d in enumerate(L_tail) if idx != i]
        T_ = ((a+1, b), L_head_ + L_tail_)
        # print(T_)
        val_ = mrTupValue(T_)
        if (val_ == val_goal):
            # i seems to always be zero
            return (i, i + len(L_head_))
    return None
#

In [14]:
for i in range(8,100):
    label = collatzPath(i)
    if countZeros(label) == 2:
        print (i, label)
#

12 1100111
13 0110111
24 11100111
26 10110111
48 111100111
52 110110111
53 011110111
96 1111100111


# 1 Composition Options

$ \Large{ \frac{2^{0}}{3^{0}} }$

$ \Large{ \frac{2^{2}}{3^{1}} - (   \frac{2^{0}}{3^{1}} ) }$

$ \Large{ \frac{2^{4}}{3^{2}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} )  }$

$ \Large{ \frac{2^{6}}{3^{3}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} )  }$

$ \Large{ \frac{2^{8}}{3^{4}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} )  }$

$ \Large{ \frac{2^{10}}{3^{5}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} + \frac{2^{8}}{3^{5}} )  }$

$ \Large{ \frac{2^{12}}{3^{6}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} + \frac{2^{8}}{3^{5}} + \frac{2^{10}}{3^{6}} )  }$

$ \Large{ \frac{2^{14}}{3^{7}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} + \frac{2^{8}}{3^{5}} + \frac{2^{10}}{3^{6}} + \frac{2^{12}}{3^{7}} )  }$

In [107]:
label = ""
for i in range(8):
    T = mrTupFromPath(label)
    print(T)
    label = label + "01"
#

((0, 0), [])
((2, -1), [(0, -1)])
((4, -2), [(0, -1), (2, -2)])
((6, -3), [(0, -1), (2, -2), (4, -3)])
((8, -4), [(0, -1), (2, -2), (4, -3), (6, -4)])
((10, -5), [(0, -1), (2, -2), (4, -3), (6, -4), (8, -5)])
((12, -6), [(0, -1), (2, -2), (4, -3), (6, -4), (8, -5), (10, -6)])
((14, -7), [(0, -1), (2, -2), (4, -3), (6, -4), (8, -5), (10, -6), (12, -7)])


In [108]:
label = ""
for i in range(8):
    T = mrTupFromPath(label)
    print(mrTupToLaTex(T))
    label = label + "01"
#

$ \frac{2^{0}}{3^{0}} $
$ \frac{2^{2}}{3^{1}} - (   \frac{2^{0}}{3^{1}} ) $
$ \frac{2^{4}}{3^{2}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} ) $
$ \frac{2^{6}}{3^{3}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} ) $
$ \frac{2^{8}}{3^{4}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} ) $
$ \frac{2^{10}}{3^{5}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} + \frac{2^{8}}{3^{5}} ) $
$ \frac{2^{12}}{3^{6}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} + \frac{2^{8}}{3^{5}} + \frac{2^{10}}{3^{6}} ) $
$ \frac{2^{14}}{3^{7}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{4}}{3^{3}} + \frac{2^{6}}{3^{4}} + \frac{2^{8}}{3^{5}} + \frac{2^{10}}{3^{6}} + \frac{2^{12}}{3^{7}} ) $


# 2 Composition Options

$ \Large{ \frac{2^{1}}{3^{0}} } $

$ \Large{ \frac{2^{3}}{3^{1}} - (   \frac{2^{1}}{3^{1}} ) } $

$ \Large{ \frac{2^{5}}{3^{2}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} ) } $

$ \Large{ \frac{2^{7}}{3^{3}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} ) } $

$ \Large{ \frac{2^{9}}{3^{4}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} ) } $

$ \Large{ \frac{2^{11}}{3^{5}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} + \frac{2^{9}}{3^{5}} ) } $

$ \Large{ \frac{2^{13}}{3^{6}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} + \frac{2^{9}}{3^{5}} + \frac{2^{11}}{3^{6}} ) } $

$ \Large{ \frac{2^{15}}{3^{7}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} + \frac{2^{9}}{3^{5}} + \frac{2^{11}}{3^{6}} + \frac{2^{13}}{3^{7}} ) } $

In [109]:
label = "1"
for i in range(8):
    T = mrTupFromPath(label)
    print(T)
    label = label + "01"
#

((1, 0), [])
((3, -1), [(1, -1)])
((5, -2), [(1, -1), (3, -2)])
((7, -3), [(1, -1), (3, -2), (5, -3)])
((9, -4), [(1, -1), (3, -2), (5, -3), (7, -4)])
((11, -5), [(1, -1), (3, -2), (5, -3), (7, -4), (9, -5)])
((13, -6), [(1, -1), (3, -2), (5, -3), (7, -4), (9, -5), (11, -6)])
((15, -7), [(1, -1), (3, -2), (5, -3), (7, -4), (9, -5), (11, -6), (13, -7)])


In [110]:
label = "1"
for i in range(8):
    T = mrTupFromPath(label)
    print(mrTupToLaTex(T))
    label = label + "01"
#

$ \frac{2^{1}}{3^{0}} $
$ \frac{2^{3}}{3^{1}} - (   \frac{2^{1}}{3^{1}} ) $
$ \frac{2^{5}}{3^{2}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} ) $
$ \frac{2^{7}}{3^{3}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} ) $
$ \frac{2^{9}}{3^{4}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} ) $
$ \frac{2^{11}}{3^{5}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} + \frac{2^{9}}{3^{5}} ) $
$ \frac{2^{13}}{3^{6}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} + \frac{2^{9}}{3^{5}} + \frac{2^{11}}{3^{6}} ) $
$ \frac{2^{15}}{3^{7}} - (   \frac{2^{1}}{3^{1}} + \frac{2^{3}}{3^{2}} + \frac{2^{5}}{3^{3}} + \frac{2^{7}}{3^{4}} + \frac{2^{9}}{3^{5}} + \frac{2^{11}}{3^{6}} + \frac{2^{13}}{3^{7}} ) $


# Basis Set: $(\frac{1}{3})^b$

At multiple '01' paddings


In [15]:
collatzPath2((1, 3))

('0', [(1, 3), (1, 1)])

In [16]:
T_3 = mrTupFromPath("0")
T_3

((1, -1), [(0, -1)])

In [17]:
T_3_01 = mrTupFromPath("001")
mrTupValue(T_3_01)

(1, 3)

In [18]:
T_3_01

((3, -2), [(0, -1), (1, -2)])

In [20]:
label = "0"
for i in range(3):
    T = mrTupFromPath(label)
    print("#### " + mrTupToLaTex(T))
    print("")
    label = label + "01"
#


#### $ \frac{2^{1}}{3^{1}} - (   \frac{2^{0}}{3^{1}} ) $

#### $ \frac{2^{3}}{3^{2}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{1}}{3^{2}} ) $

#### $ \frac{2^{5}}{3^{3}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{1}}{3^{2}} + \frac{2^{3}}{3^{3}} ) $



# 1/3 composition options given by mrTup degeneracy:


#### $ \frac{2^{1}}{3^{1}} - (   \frac{2^{0}}{3^{1}} ) $

#### $ \frac{2^{3}}{3^{2}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{1}}{3^{2}} ) $

#### $ \frac{2^{5}}{3^{3}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{1}}{3^{2}} + \frac{2^{3}}{3^{3}} ) $

In [21]:
collatzPath2((1, 9))

('010', [(1, 9), (2, 3), (1, 3), (1, 1)])

In [22]:
label = "010"
for i in range(3):
    T = mrTupFromPath(label)
    print("#### " + mrTupToLaTex(T))
    print("")
    label = label + "01"
#


#### $ \frac{2^{3}}{3^{2}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} ) $

#### $ \frac{2^{5}}{3^{3}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{3}}{3^{3}} ) $

#### $ \frac{2^{7}}{3^{4}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{3}}{3^{3}} + \frac{2^{5}}{3^{4}} ) $



# 1/9 composition options given by mrTup degeneracy:

#### $ \frac{2^{3}}{3^{2}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} ) $

#### $ \frac{2^{5}}{3^{3}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{3}}{3^{3}} ) $

#### $ \frac{2^{7}}{3^{4}} - (   \frac{2^{0}}{3^{1}} + \frac{2^{2}}{3^{2}} + \frac{2^{3}}{3^{3}} + \frac{2^{5}}{3^{4}} ) $

## Tuple constraints do not admit simple decrementing of denominator to generate $ 3\cdot\frac{1}{9} \mapsto \frac{1}{3} $

$F_{\frac{1}{3}}(T) ? \mapsto ? ((a+2, b+1), L + [(L[-1][0]+1, L[-1][1]-1))])$  e.g. start with  4/3 mutliplication of primary ...

↥↥↥ Works for 1/3 ... but in general more complicated


In [23]:
for i in range(3, 200, 1):
    T = mrTupFromValue(i)
    if len(T[1]) > 0:
        T_ = ((T[0][0] + 2, T[0][1] - 1), T[1] + [(T[1][-1][0] + 1, T[1][-1][1] -1)] )
        val_ = mrTupValue(T_)
        if val_[1] == 1:
            print(f"{i} --thirdish-> {val_[0]}")

In [24]:
mrTupFromValue(21), mrTupFromValue(7)

(((6, -1), [(0, -1)]),
 ((11, -5), [(0, -1), (1, -2), (2, -3), (4, -4), (7, -5)]))

In [25]:
# ^^^ Definitely more complicated.

In [26]:
collatzPath(21)

'011111'

In [55]:
# 0 (mod 8) non-neighborly
for i in range(8, 50, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    label_ = collatzPath(2*i+1)
    print((i, is_2np1easy, label, label_))

(8, False, '111', '010110111')
(16, False, '1111', '010100110010110111')
(24, False, '11100111', '01011100010110111')
(32, False, '11111', '0101011100010110111')
(40, False, '1110111', '0101100011110111')
(48, False, '111100111', '010100011000010100100010000101100010010000001100001110101011101100011110111')


In [201]:
for i in range(4, 200, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    label_ = collatzPath(2*i+1)
    print((i, (2*i +1)%8, None, is_2np1easy, label, label_))

(4, 1, None, False, '11', '0100010110111')
(12, 1, None, False, '1100111', '0100110010110111')
(20, 1, None, False, '110111', '010000010100100010000101100010010000001100001110101011101100011110111')
(28, 1, None, False, '1100010110111', '0100101011100010110111')
(36, 1, None, False, '110100010110111', '0100011000010100100010000101100010010000001100001110101011101100011110111')
(44, 1, None, False, '110010110111', '010011100110010110111')
(52, 1, None, False, '110110111', '01000011011100110010110111')
(60, 1, None, False, '11000011110111', '0100100010000101100010010000001100001110101011101100011110111')
(68, 1, None, False, '11010110111', '0100010000101100010010000001100001110101011101100011110111')
(76, 1, None, False, '1100110010110111', '0100110101011100010110111')
(84, 1, None, False, '11011111', '010000000101110110110110010110111')
(92, 1, None, False, '1100011110111', '010010110010011100110010110111')
(100, 1, None, False, '110100110010110111', '01000111111111')
(108, 1, None, Fal

In [48]:
# 2 (mod 8) neighborly
for i in range(2, 50, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(2, True, '1')
(10, True, '10111')
(18, True, '10100010110111')
(26, True, '10110111')
(34, True, '1010110111')
(42, True, '1011111')


In [57]:
# 6 (mod 8) neighborly
for i in range(6, 50, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(6, True, '100111')
(14, True, '100010110111')
(22, True, '10010110111')
(30, True, '1000011110111')
(38, True, '100110010110111')
(46, True, '100011110111')


In [28]:
# 1 (mod 4) mixed ... what started me looking at 8n+b numbers
for i in range(1, 44, 4):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(1, True, '')
(5, False, '0111')
(9, True, '0100010110111')
(13, False, '0110111')
(17, True, '010110111')
(21, False, '011111')
(25, True, '0100110010110111')
(29, False, '0110010110111')
(33, True, '010100110010110111')
(37, False, '011100010110111')
(41, True, '010000010100100010000101100010010000001100001110101011101100011110111')


In [197]:
# 1 (mod 8) neighborly
for i in range(9, 100, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    label_ = collatzPath(2*i+1)
    print((i, (2*i +1)%8, ((i-1)//2)%8, is_2np1easy, label, label_))

(9, 3, 4, True, '0100010110111', '00110010110111')
(17, 3, 0, True, '010110111', '0011110111')
(25, 3, 4, True, '0100110010110111', '00110110010110111')
(33, 3, 0, True, '010100110010110111', '0011100110010110111')
(41, 3, 4, True, '010000010100100010000101100010010000001100001110101011101100011110111', '0011000010100100010000101100010010000001100001110101011101100011110111')
(49, 3, 0, True, '01011100010110111', '001111100010110111')
(57, 3, 4, True, '0100101011100010110111', '00110101011100010110111')
(65, 3, 0, True, '0101011100010110111', '00111011100010110111')
(73, 3, 4, True, '0100011000010100100010000101100010010000001100001110101011101100011110111', '00110011000010100100010000101100010010000001100001110101011101100011110111')
(81, 3, 0, True, '0101100011110111', '00111100011110111')
(89, 3, 4, True, '010011100110010110111', '0011011100110010110111')
(97, 3, 0, True, '010100011000010100100010000101100010010000001100001110101011101100011110111', '00111000110000101001000100001011

In [60]:
# 5 (mod 8) non-neighborly
for i in range(5, 100, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    label_ = collatzPath(2*i+1)
    print((i, is_2np1easy, label, label_))

(5, False, '0111', '0010110111')
(13, False, '0110111', '0010000010100100010000101100010010000001100001110101011101100011110111')
(21, False, '011111', '00101011100010110111')
(29, False, '0110010110111', '0010011100110010110111')
(37, False, '011100010110111', '00101111111')
(45, False, '011010110111', '00100010000101100010010000001100001110101011101100011110111')
(53, False, '011110111', '0010100100010000101100010010000001100001110101011101100011110111')
(61, False, '01100011110111', '0010010110010011100110010110111')
(69, False, '01110110111', '0010110010011100110010110111')
(77, False, '0110110010110111', '0010000101100010010000001100001110101011101100011110111')
(85, False, '01111111', '0010101011010000010100100010000101100010010000001100001110101011101100011110111')
(93, False, '0110011110111', '001001100011011100110010110111')


In [196]:
# 3 (mod 8) non-neighborly
for i in range(3, 100, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    label_ = collatzPath(2*i+1)
    print((i, (2*i +1)%8, ((i-1)//2)%8, is_2np1easy, label, label_))

(3, 7, 1, False, '00111', '00010110111')
(11, 7, 5, True, '0010110111', '00011110111')
(19, 7, 1, False, '00110010110111', '00010011100110010110111')
(27, 7, 5, True, '0010000010100100010000101100010010000001100001110101011101100011110111', '00011000010100100010000101100010010000001100001110101011101100011110111')
(35, 7, 1, False, '0011110111', '00010100100010000101100010010000001100001110101011101100011110111')
(43, 7, 5, True, '00101011100010110111', '000111011100010110111')
(51, 7, 1, False, '00110110010110111', '00010000101100010010000001100001110101011101100011110111')
(59, 7, 5, True, '0010011100110010110111', '00011011100110010110111')
(67, 7, 1, False, '0011100110010110111', '0001011100101011100010110111')
(75, 7, 5, True, '00101111111', '000111111111')
(83, 7, 1, False, '0011000010100100010000101100010010000001100001110101011101100011110111', '00010010000001100001110101011101100011110111')
(91, 7, 5, True, '00100010000101100010010000001100001110101011101100011110111', '000110

In [50]:
for i in range(7, 100, 8):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(7, True, '00010110111')
(15, False, '000011110111')
(23, False, '00011110111')
(31, True, '0000010100100010000101100010010000001100001110101011101100011110111')
(39, True, '00010011100110010110111')
(47, True, '000010100100010000101100010010000001100001110101011101100011110111')
(55, False, '00011000010100100010000101100010010000001100001110101011101100011110111')
(63, False, '00000011100100010000101100010010000001100001110101011101100011110111')
(71, True, '00010100100010000101100010010000001100001110101011101100011110111')
(79, False, '000011011100110010110111')
(87, False, '000111011100010110111')
(95, False, '0000011100100010000101100010010000001100001110101011101100011110111')


## 0[11]+ labels are $n' = 4n+1$ numbers, $n_0 = 0$
We get one of these numbers every other (even generation)

In [31]:
label = '01'
for i in range(10):
    n_d = mrTupValue(mrTupFromPath(label))
    if n_d[1] == 1:
        print((n_d[0], label))
    label = label + "1"
#

(1, '01')
(5, '0111')
(21, '011111')
(85, '01111111')
(341, '0111111111')


## 10[11]+ labels are $n' = 4n+2$ power-series numbers, $n_0 = 0$
In odd generations

In [32]:
label = '101'
for i in range(10):
    n_d = mrTupValue(mrTupFromPath(label))
    if n_d[1] == 1:
        is_2np1easy  = checkTupEasy(label)
        print((n_d[0], is_2np1easy, label))
    label = label + "1"
#

(2, True, '101')
(10, True, '10111')
(42, True, '1011111')
(170, True, '101111111')
(682, True, '10111111111')


In [33]:
for i in range(2, 44, 4):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(2, True, '1')
(6, True, '100111')
(10, True, '10111')
(14, True, '100010110111')
(18, True, '10100010110111')
(22, True, '10010110111')
(26, True, '10110111')
(30, True, '1000011110111')
(34, True, '1010110111')
(38, True, '100110010110111')
(42, True, '1011111')


In [34]:
for i in range(4, 44, 4):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(4, False, '11')
(8, False, '111')
(12, False, '1100111')
(16, False, '1111')
(20, False, '110111')
(24, False, '11100111')
(28, False, '1100010110111')
(32, False, '11111')
(36, False, '110100010110111')
(40, False, '1110111')


In [35]:
val = 3
for i in range(10):
    label = collatzPath(val)
    is_2np1easy  = checkTupEasy(label)
    parent_2np1 = (val -1)//2
    print((val, is_2np1easy, checkTupEasy(collatzPath(parent_2np1)),  label))
    val = 4*val + 3
#

(3, False, True, '00111')
(15, False, True, '000011110111')
(63, False, True, '00000011100100010000101100010010000001100001110101011101100011110111')
(255, False, True, '00000000111110110110110010110111')
(1023, False, True, '000000000011101111110110101011100010110111')
(4095, False, True, '00000000000011110000001000110101010010110100100111100101011001010111010000000101110110110110010110111')
(16383, False, True, '0000000000000011100001001100111001001111011110000011000111011001010111010000000101110110110110010110111')
(65535, False, True, '00000000000000001111110110011100100000010011101100111001010101111111110100110010110111')
(262143, False, True, '0000000000000000001110100000100100001111010110110000001001100110010001110100110101010110100001100101110111000111010001100110011101111110010110111')
(1048575, False, True, '000000000000000000001111010010111100110101011010100101011110000000110011101111101110101011001011100101011100010110111')


In [36]:
# n' = ((4n+3)-1)/2 -> 2n+1
# 2(2n+1) + 1 -> 4n+3 !

# All $4n+3$ Numbers are *'$2n+1$ non-neighborly'*

### Because $2(2n+1) + 1 = 4n+3$

So not just the numbers in the $n' = 4n+3$ power series

In [68]:
for i in range(3, 100, 4):
    label = collatzPath(i)
    T = mrTupFromPath(label)
    T_ = F_0(T)
    val_ = mrTupValue(T_)
    
    print((i, val_[0] % 8, label, val_))

(3, 1, '00111', (17, 27))
(7, 7, '00010110111', (1007, 729))
(11, 1, '0010110111', (625, 243))
(15, 7, '000011110111', (2743, 729))
(19, 1, '00110010110111', (8785, 2187))
(23, 5, '00011110111', (1493, 243))
(27, 3, '0010000010100100010000101100010010000001100001110101011101100011110111', (593129465116011091795, 109418989131512359209))
(31, 7, '0000010100100010000101100010010000001100001110101011101100011110111', (81739724051411966975, 12157665459056928801))
(35, 3, '0011110111', (787, 81))
(39, 7, '00010011100110010110111', (3948983, 531441))
(43, 3, '00101011100010110111', (441955, 59049))
(47, 5, '000010100100010000101100010010000001100001110101011101100011110111', (42896139602215471621, 4052555153018976267))
(51, 3, '00110110010110111', (72467, 6561))
(55, 7, '00011000010100100010000101100010010000001100001110101011101100011110111', (1295677919363534542799, 109418989131512359209))
(59, 1, '0010011100110010110111', (2063065, 177147))
(63, 7, '0000001110010001000010110001001000000110

In [69]:
for i in range(3, 100, 4):
    label = collatzPath(i)
    T = mrTupFromPath(label)
    T_ = F_1(T)
    val_ = mrTupValue(T_)
    
    print((i, val_[0] % 8, label, val_))

(3, 3, '00111', (59, 9))
(7, 5, '00010110111', (3749, 243))
(11, 3, '0010110111', (1915, 81))
(15, 5, '000011110111', (7741, 243))
(19, 3, '00110010110111', (30235, 729))
(23, 7, '00011110111', (3911, 81))
(27, 1, '0010000010100100010000101100010010000001100001110101011101100011110111', (2165362522901022536305, 36472996377170786403))
(31, 5, '0000010100100010000101100010010000001100001110101011101100011110111', (273203162333264677205, 4052555153018976267))
(35, 1, '0011110111', (1969, 27))
(39, 5, '00010011100110010110111', (15297341, 177147))
(43, 1, '00101011100010110111', (1894945, 19683))
(47, 7, '000010100100010000101100010010000001100001110101011101100011110111', (137277007025468834647, 1350851717672992089))
(51, 1, '00110110010110111', (242609, 2187))
(55, 5, '00011000010100100010000101100010010000001100001110101011101100011110111', (4367198042179215859013, 36472996377170786403))
(59, 3, '0010011100110010110111', (7678195, 59049))
(63, 5, '000000111001000100001011000100100000011

In [66]:
for i in range(3, 100, 4):
    label = collatzPath(i)
    T = mrTupFromPath(label)
    T_ = F_1(T)
    val_ = mrTupValue(T_)
    if (val_[0] % 8) in [0,2,4,6]:
        print((i, val_[0] % 8, label, val_))

In [37]:
collatzPath2((1, 27))

('00110', [(1, 27), (5, 9), (4, 3), (2, 3), (1, 3), (1, 1)])

In [38]:
mrTupFromPath('00110')

((5, -3), [(0, -1), (1, -2), (4, -3)])

In [39]:
mrTupValue(mrTupFromPath('00000011100100010000101100010010000001100001110101011101100011110111'))

(63, 1)

In [40]:
# 1/3 degenerate to same number of zeros as 1/27 ...
#   Note only difference is power of two of last term.
mrTupFromPath("00101")

((5, -3), [(0, -1), (1, -2), (3, -3)])

In [41]:
collatzPath2((1, 81))

('010111010',
 [(1, 81),
  (14, 27),
  (7, 27),
  (8, 9),
  (4, 9),
  (2, 9),
  (1, 9),
  (2, 3),
  (1, 3),
  (1, 1)])

In [42]:
collatzPath2((1, 243))

('001001110',
 [(1, 243),
  (41, 81),
  (34, 27),
  (17, 27),
  (13, 9),
  (8, 3),
  (4, 3),
  (2, 3),
  (1, 3),
  (1, 1)])

In [43]:
collatzPath2((1, 3**6))

('01000011110',
 [(1, 729),
  (122, 243),
  (61, 243),
  (71, 81),
  (49, 27),
  (29, 9),
  (16, 3),
  (8, 3),
  (4, 3),
  (2, 3),
  (1, 3),
  (1, 1)])

In [44]:
collatzPath2((1, 3**7))

('001111010100011',
 [(1, 2187),
  (365, 729),
  (304, 243),
  (152, 243),
  (76, 243),
  (38, 243),
  (19, 243),
  (50, 81),
  (25, 81),
  (26, 27),
  (13, 27),
  (11, 9),
  (7, 3),
  (4, 1),
  (2, 1),
  (1, 1)])

In [45]:
collatzPath2((1, 3**8))

('010100011100011',
 [(1, 6561),
  (1094, 2187),
  (547, 2187),
  (638, 729),
  (319, 729),
  (281, 243),
  (181, 81),
  (104, 27),
  (52, 27),
  (26, 27),
  (13, 27),
  (11, 9),
  (7, 3),
  (4, 1),
  (2, 1),
  (1, 1)])

# (mod 8) manipulations

Using lattice transformations to bend labels to our will

In [92]:
# 0 (mod 8) 
#  Prepending "1" puts us right back where we started, so what happens with a 0?
for i in range(8, 50, 8):
    label = collatzPath(i)
    label_ = "0" + label
    T_ = mrTupFromPath(label_)
    val_ = mrTupValue(T_)
    print((i, val_, label_))

(8, (5, 1), '0111')
(16, (31, 3), '01111')
(24, (47, 3), '011100111')
(32, (21, 1), '011111')
(40, (79, 3), '01110111')
(48, (95, 3), '0111100111')
The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


In [104]:
for i in range(8, 50, 8):
    label = collatzPath(i)
    label_ = label[0:-3] + "00111"
    T_ = mrTupFromPath(label_)
    val_ = mrTupValue(T_)
    print((i, val_, label_))

(8, (3, 1), '00111')
(16, (6, 1), '100111')
(24, (56, 9), '1110000111')
(32, (12, 1), '1100111')
(40, (40, 3), '111000111')
(48, (112, 9), '11110000111')


# (mod 12)

In [70]:
# 0 (mod 6) neighborly
for i in range(6, 50, 6):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(6, True, '100111')
(12, False, '1100111')
(18, True, '10100010110111')
(24, False, '11100111')
(30, True, '1000011110111')
(36, False, '110100010110111')
(42, True, '1011111')
(48, False, '111100111')


In [73]:
# 0 (mod 12)
for i in range(12, 50, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(12, False, '1100111')
(24, False, '11100111')
(36, False, '110100010110111')
(48, False, '111100111')


In [76]:
# 1 (mod 12)
for i in range(1, 50, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(1, True, '')
(13, False, '0110111')
(25, True, '0100110010110111')
(37, False, '011100010110111')
(49, True, '01011100010110111')


In [80]:
# 1 (mod 24)
for i in range(1, 100, 24):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(1, True, '')
(25, True, '0100110010110111')
(49, True, '01011100010110111')
(73, True, '0100011000010100100010000101100010010000001100001110101011101100011110111')
(97, True, '010100011000010100100010000101100010010000001100001110101011101100011110111')


In [81]:
# 13 (mod 24)
for i in range(13, 100, 24):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(13, False, '0110111')
(37, False, '011100010110111')
(61, False, '01100011110111')
(85, False, '01111111')


In [82]:
# 2 (mod 12)
for i in range(2, 50, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(2, True, '1')
(14, True, '100010110111')
(26, True, '10110111')
(38, True, '100110010110111')


In [83]:
# 3 (mod 12)
for i in range(3, 50, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(3, False, '00111')
(15, False, '000011110111')
(27, True, '0010000010100100010000101100010010000001100001110101011101100011110111')
(39, True, '00010011100110010110111')


In [84]:
# 4 (mod 12)
for i in range(4, 50, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(4, False, '11')
(16, False, '1111')
(28, False, '1100010110111')
(40, False, '1110111')


In [85]:
# 5 (mod 12)
for i in range(5, 50, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(5, False, '0111')
(17, True, '010110111')
(29, False, '0110010110111')
(41, True, '010000010100100010000101100010010000001100001110101011101100011110111')


In [87]:
# 5 (mod 24)
for i in range(5, 100, 24):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(5, False, '0111')
(29, False, '0110010110111')
(53, False, '011110111')
(77, False, '0110110010110111')


In [88]:
# 17 (mod 24)
for i in range(17, 100, 24):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(17, True, '010110111')
(41, True, '010000010100100010000101100010010000001100001110101011101100011110111')
(65, True, '0101011100010110111')
(89, True, '010011100110010110111')


In [75]:
# 6 (mod 12)
for i in range(6, 50, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(6, True, '100111')
(18, True, '10100010110111')
(30, True, '1000011110111')
(42, True, '1011111')


# (mod 12) is a FAIL
Below we can see that 7 (mod 12) numbers are not consistent and are not going to divide up cleanly into some higher base set

In [91]:
# (mod 12) sort of falls apart here ...
# 7 (mod 12)
for i in range(7, 100, 12):
    label = collatzPath(i)
    is_2np1easy  = checkTupEasy(label)
    print((i, is_2np1easy, label))

(7, True, '00010110111')
(19, False, '00110010110111')
(31, True, '0000010100100010000101100010010000001100001110101011101100011110111')
(43, True, '00101011100010110111')
(55, False, '00011000010100100010000101100010010000001100001110101011101100011110111')
(67, False, '0011100110010110111')
(79, False, '000011011100110010110111')
(91, True, '00100010000101100010010000001100001110101011101100011110111')


# $7 \pmod{8} $ Numbers

Can we show all $7 \pmod{8} $ are in the lattice using a density argument

In [113]:
gen_7_mod8_neighborly = {}
gen_7_mod8_nonneighborly = {}

for i in range(1000):
    n = 8*i + 7
    label = collatzPath(n)
    a = len(label)
    if checkTupEasy(label):
        if a not in gen_7_mod8_neighborly:
            gen_7_mod8_neighborly[a] = []
        gen_7_mod8_neighborly[a].append((n, label))
    else:
        if a not in gen_7_mod8_nonneighborly:
            gen_7_mod8_nonneighborly[a] = []
        gen_7_mod8_nonneighborly[a].append((n, label))
            

In [114]:
len(gen_7_mod8_neighborly), len(gen_7_mod8_nonneighborly)

(115, 111)

In [117]:
sorted(gen_7_mod8_neighborly.keys())

[11,
 21,
 22,
 23,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 130,
 131,
 132,
 137,
 143,
 149,
 150,
 154,
 162]

In [118]:
sorted(gen_7_mod8_nonneighborly.keys())

[11,
 12,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 121,
 122,
 124,
 127,
 132,
 133,
 138,
 151]

In [126]:
# Neighborly
gen_7_mod8_gen_pairs = {}

for i in range(1000):
    n = 8*i + 7
    label = collatzPath(n)
    a = len(label)
    if checkTupEasy(label):
        n_ = 2*n + 1
        label_ = collatzPath(n_)
        if a not in gen_7_mod8_gen_pairs:
            gen_7_mod8_gen_pairs[a] = {}
        a_ = len(label_)
        if a_ not in gen_7_mod8_gen_pairs[a]:
            gen_7_mod8_gen_pairs[a][a_] = set([])
        gen_7_mod8_gen_pairs[a][a_].add(n_)
    else:
        n_ = (n - 1)//2
        label_ = collatzPath(n_)
        a_ = len(label_)
        if a_ not in gen_7_mod8_gen_pairs:
            gen_7_mod8_gen_pairs[a_] = {}
        if a not in gen_7_mod8_gen_pairs[a_]:
            gen_7_mod8_gen_pairs[a_][a] = set([])
        gen_7_mod8_gen_pairs[a_][a].add(n)
        
            

In [123]:
gen_7_mod8_gen_pairs

{11: {12: {15, 151}},
 10: {11: {23}},
 67: {68: {63, 1591, 4919, 4983, 5207, 5239, 14159, 14303, 14655}},
 23: {24: {79, 847, 2199, 7639}},
 66: {67: {95, 2623, 7063, 7263, 7351, 7831, 8799}},
 70: {71: {55, 4191, 4303, 4631, 4663, 4695, 12559, 12623, 14095}},
 65: {66: {143, 1303, 3583, 3935, 10623, 10895, 11871, 13199}},
 20: {21: {87, 951, 2679}},
 56: {57: {207,
   1679,
   1879,
   5071,
   5215,
   5327,
   15167,
   15311,
   15567,
   15615,
   15679,
   15711,
   15743}},
 45: {46: {223, 1847, 5399, 5439, 5495, 5583, 5591, 5599, 6039}},
 22: {23: {119, 1111, 1143, 1167, 1271}},
 31: {32: {247, 255, 727, 735, 6775, 7375, 7439, 8015, 8031}},
 28: {29: {271, 279, 863, 911, 8463, 9055}},
 36: {37: {319, 855, 887, 2583, 2647, 2655, 2895, 7767, 7775, 7927, 7951}},
 44: {45: {335, 2775, 8159, 8287, 8399, 8671}},
 52: {53: {351, 2871, 8591, 8767, 8847, 9567, 10559, 10575}},
 59: {60: {183,
   4535,
   4671,
   13711,
   13791,
   13919,
   13967,
   14591,
   15071,
   15327,
   1545

In [129]:
# Non-Neighborly
gen_7_mod8_gen_pairs_nn = {}

for i in range(1000):
    n = 8*i + 7
    label = collatzPath(n)
    a = len(label)
    if checkTupEasy(label):
        n_ = (n - 1)//2
        label_ = collatzPath(n_)
        if a not in gen_7_mod8_gen_pairs_nn:
            gen_7_mod8_gen_pairs_nn[a] = {}
        a_ = len(label_)
        if a_ not in gen_7_mod8_gen_pairs_nn[a]:
            gen_7_mod8_gen_pairs_nn[a][a_] = set([])
        gen_7_mod8_gen_pairs_nn[a][a_].add(n_)
    else:
        n_ = 2*n + 1
        label_ = collatzPath(n_)
        a_ = len(label_)
        if a_ not in gen_7_mod8_gen_pairs_nn:
            gen_7_mod8_gen_pairs_nn[a_] = {}
        if a not in gen_7_mod8_gen_pairs_nn[a_]:
            gen_7_mod8_gen_pairs_nn[a_][a] = set([])
        gen_7_mod8_gen_pairs_nn[a_][a].add(n)

            

In [130]:
gen_7_mod8_gen_pairs_nn

{11: {5: {3}},
 67: {12: {15}, 34: {3575}, 47: {3663}, 39: {3539}},
 66: {11: {23},
  92: {655, 5967},
  38: {1815},
  24: {2199},
  33: {5343},
  65: {5431},
  46: {5495},
  30: {5527},
  100: {6015}},
 23: {14: {19}, 27: {211}},
 45: {71: {55},
  44: {1359, 1399, 5007},
  82: {1395},
  101: {4055},
  98: {4559, 4599}},
 31: {68: {63},
  60: {183},
  30: {2007},
  84: {1843, 1859},
  92: {2003},
  46: {6039}},
 65: {10: {35},
  64: {895},
  91: {983, 2967},
  37: {2655, 2723},
  23: {3299},
  121: {7999}},
 36: {24: {79}, 19: {663}, 32: {723}, 27: {1943}, 62: {1987}, 35: {7119}},
 52: {21: {87},
  51: {2191},
  78: {2391, 6647},
  94: {2639},
  19: {2147},
  62: {2211},
  67: {2643},
  35: {6415},
  81: {6423},
  48: {7231}},
 30: {67: {95}, 59: {275}, 86: {835}, 29: {2707, 3011}},
 56: {17: {51},
  28: {419, 3891},
  66: {1303, 3935},
  23: {1267},
  36: {1331},
  112: {3791, 3903},
  55: {3827, 3919, 3927, 4239},
  25: {4247}},
 35: {23: {119},
  64: {991},
  42: {1039},
  34: {963}

In [135]:
# What is our 8k+7 neighborly generation rate per generation?
accum = 0
for a in sorted(gen_7_mod8_gen_pairs.keys()):
    j = len(gen_7_mod8_gen_pairs[a][a+1])
    accum +=j
    print((a, j, accum/(2**(a-3))))


(10, 1, 0.0078125)
(11, 2, 0.01171875)
(16, 1, 0.00048828125)
(17, 1, 0.00030517578125)
(18, 1, 0.00018310546875)
(19, 1, 0.0001068115234375)
(20, 3, 7.62939453125e-05)
(21, 3, 4.9591064453125e-05)
(22, 5, 3.4332275390625e-05)
(23, 4, 2.09808349609375e-05)
(24, 2, 1.1444091796875e-05)
(25, 3, 6.4373016357421875e-06)
(26, 8, 4.172325134277344e-06)
(27, 10, 2.682209014892578e-06)
(28, 6, 1.519918441772461e-06)
(29, 8, 8.791685104370117e-07)
(30, 10, 5.140900611877441e-07)
(31, 9, 2.905726432800293e-07)
(32, 10, 1.6391277313232422e-07)
(33, 7, 8.847564458847046e-08)
(34, 14, 5.075708031654358e-08)
(35, 13, 2.8405338525772095e-08)
(36, 11, 1.548323780298233e-08)
(37, 9, 8.265487849712372e-09)
(38, 9, 4.394678398966789e-09)
(39, 10, 2.342858351767063e-09)
(40, 15, 1.280568540096283e-09)
(41, 11, 6.803020369261503e-10)
(42, 10, 3.583409124985337e-10)
(43, 12, 1.90084392670542e-10)
(44, 6, 9.777068044058979e-11)
(45, 9, 5.093170329928398e-11)
(46, 14, 2.7057467377744615e-11)
(47, 8, 1.3983481

### ^^^ Clearly the wrong way to look at this

For one by incrementing over the integers instead of complete generations we have no idea what our fill factor is.

Lets just count per-genration the number of $7 \pmod{8}$ numbers we have:

In [163]:
# This is slow ... computation up to generation 26 below
for a in range(1,20,1):
    """ Count the number of 7 (mod 8) numbers found per generation.
    """
    accum = [0]*8
    int_count = 0
    for label in generationLabels(a):
        T = mrTupFromPath(label)
        val = mrTupValue(T)
        if val[1] == 1:
            int_count += 1
            accum[(val[0] % 8)] += 1
            if (val[0] % 8) == 7:
                print((7, a, val, label, T))
            #if (val[0] % 8) == 3:
            #    print((3, a, val, label, T))
    print((a, accum, int_count))
#
comment = """
(10, [10, 2, 3, 2, 4, 4, 1, 0], 26)
(11, [14, 1, 6, 3, 4, 4, 2, 2], 36)
(12, [18, 2, 5, 2, 8, 8, 5, 2], 50)
(13, [26, 2, 10, 3, 10, 10, 4, 2], 67)
(14, [36, 3, 12, 3, 14, 14, 5, 2], 89)
(15, [50, 5, 17, 4, 17, 17, 5, 2], 117)
(16, [67, 9, 22, 7, 22, 22, 6, 2], 157)
(17, [89, 9, 31, 11, 28, 28, 9, 3], 208)
(18, [117, 14, 37, 12, 40, 40, 14, 3], 277)
(19, [157, 18, 54, 17, 51, 51, 15, 4], 367)
(20, [208, 23, 69, 25, 69, 69, 21, 4], 488)
(21, [277, 34, 92, 28, 90, 90, 29, 9], 649)
(22, [367, 44, 124, 47, 121, 121, 37, 8], 869)
(23, [488, 53, 165, 53, 161, 161, 55, 18], 1154)
(24, [649, 74, 214, 71, 220, 220, 71, 15], 1534)
(25, [869, 100, 294, 93, 285, 285, 86, 27], 2039)
(26, [1154, 135, 385, 134, 380, 380, 120, 33], 2721)
"""

(1, [0, 0, 1, 0, 0, 0, 0, 0], 1)
(2, [0, 1, 0, 0, 1, 0, 0, 0], 2)
(3, [1, 0, 1, 0, 0, 0, 0, 0], 2)
(4, [1, 1, 0, 0, 1, 1, 0, 0], 4)
(5, [2, 0, 2, 1, 0, 0, 0, 0], 5)
(6, [2, 1, 0, 0, 2, 2, 1, 0], 8)
(7, [4, 0, 3, 1, 1, 1, 0, 0], 10)
(8, [5, 1, 1, 0, 3, 3, 1, 0], 14)
(9, [8, 1, 4, 1, 2, 2, 0, 0], 18)
(10, [10, 2, 3, 2, 4, 4, 1, 0], 26)
(7, 11, (23, 1), '00011110111', ((11, -4), [(0, -1), (1, -2), (2, -3), (7, -4)]))
(7, 11, (7, 1), '00010110111', ((11, -5), [(0, -1), (1, -2), (2, -3), (4, -4), (7, -5)]))
(11, [14, 1, 6, 3, 4, 4, 2, 2], 36)
(7, 12, (151, 1), '000111111111', ((12, -3), [(0, -1), (1, -2), (2, -3)]))
(7, 12, (15, 1), '000011110111', ((12, -5), [(0, -1), (1, -2), (2, -3), (3, -4), (8, -5)]))
(12, [18, 2, 5, 2, 8, 8, 5, 2], 50)
(7, 13, (23, 1), '0001111011101', ((13, -5), [(0, -1), (1, -2), (2, -3), (7, -4), (11, -5)]))
(7, 13, (7, 1), '0001011011101', ((13, -6), [(0, -1), (1, -2), (2, -3), (4, -4), (7, -5), (11, -6)]))
(13, [26, 2, 10, 3, 10, 10, 4, 2], 67)
(7, 14, (151, 1), 

In [ ]:
1154, 385, 380, 120

### Sigh, so $7 \pmod{8}$ numbers are our slowest growing subset of $\pmod{8}$ numbers
They are not 12% of our numbers, more in the 1% to 2% of numbers per generation

How can I turn this into a good thing?

Do they always have to be in the bottom 1/8th of the lattice? **Yes**, 7 (mod 8) number labels always start with at least 3 zeros "000", add the required "111" suffix for integers and the earliest one could get a 7 (mod 8) number is generation 6 ... but we do not see one until generation 11.

We might be able to show how to generate all $7 \pmod{8}$ from 7, 23, 15, 151 because these are the first four $7 \pmod{8}$ numbers from the first 2 generations that give these numbers and we know the $7 \pmod{8}$ set is closed.

- $(11, (23, 1), '00011110111', ((11, -4), [(0, -1), (1, -2), (2, -3), (7, -4)]))$
- $(11, (7, 1), '00010110111', ((11, -5), [(0, -1), (1, -2), (2, -3), (4, -4), (7, -5)]))$
- $(12, (151, 1), '000111111111', ((12, -3), [(0, -1), (1, -2), (2, -3)]))$
- $(12, (15, 1), '000011110111', ((12, -5), [(0, -1), (1, -2), (2, -3), (3, -4), (8, -5)]))$


In [156]:
for i in range(100000):
    n = 8*i + 7
    label = collatzPath(n)
    if label[0:3] != "000":
        print((n, label))

In [159]:
collatzPath(47)

'000010100100010000101100010010000001100001110101011101100011110111'

# The 5 (mod 8) numbers are 1:1 with the 4(mod 8) integers of each generation

And they are alway related by exactly +1.

So we cannot have a 5 (mod 8) or 4(mod 8) integer without the other.

The systematic and predictable generation of $4 \pmod{8}$ and $5 \pmod{8}$ numbers in every generation of the Mixed Radix (MR) lattice points to the most fundamental **structural properties** of the Collatz map's inverse and how the MR functions ($F_0$ and $F_1$) operate on parity.

The feature causing this pattern is the **Parity Constraints imposed by the $F_0$ and $F_1$ Generators and the Collatz Inverse Map.**

---

### 1. Understanding the MR Lattice Generation

The MR lattice starts at $n=1$ (the initial tuple $T=(\mathbf{0, 0})$). Each "generation" corresponds to increasing the length of the minimal label by one step, or more generally, numbers with similar minimal path properties.

The generation functions are:
* $F_1(n) = \mathbf{2n}$ (Label $\mathbf{'1'} L(n)$)
* $F_0(n) = \frac{\mathbf{2n-1}}{3}$ (Label $\mathbf{'0'} L(n)$, if $n \equiv 2 \pmod{3}$)

### 2. Analysis of the Generated Parity

We look at the starting number $n$ and the resulting numbers' modulo 8 class after applying the $F_1$ and $F_0$ operations.

#### A. Generation of $4 \pmod{8}$ Numbers

A number $n'$ is $4 \pmod{8}$ if $n' = 8k + 4$. This is equivalent to $n' = 4(2k+1)$, meaning $n'$ is $\mathbf{4 \times \text{Odd}}$.

The only generator that multiplies a number by 2 is $\mathbf{F_1}$.

| Input $n$ (Previous Generation) | $F_1(n) = 2n$ | $2n \pmod{8}$ Result |
| :---: | :---: | :---: |
| $n \equiv 1 \pmod{4}$ (Odd) | $2(4k+1) = 8k+2$ | $\mathbf{2 \pmod{8}}$ |
| $n \equiv 3 \pmod{4}$ (Odd) | $2(4k+3) = 8k+6$ | $\mathbf{6 \pmod{8}}$ |
| **$n \equiv 2 \pmod{4}$ (Odd $\to 2 \times \text{Odd}$) ** | $2n = 2(2m) = 4m$ | **$4 \pmod{8}$ or $0 \pmod{8}$** |
| $n \equiv 4 \pmod{4}$ | $2(4m) = 8m$ | $\mathbf{0 \pmod{8}}$ |

The key insight is that the generator $\mathbf{F_1(n) = 2n}$ produces numbers whose parity depends on $n \pmod{4}$. For the lattice to produce $4 \pmod{8}$ numbers in **every generation**, it must always have a supply of **odd** numbers from the previous generation $n$ such that $\mathbf{2n}$ is a $4 \pmod{8}$ number, which is impossible.

Instead, the **structural feature is a repeated application of $\mathbf{F_1}$ on $2 \pmod{4}$ numbers**:

$$
\dots \xrightarrow{F_1} \mathbf{n_{i-1}} \xrightarrow{F_1} \mathbf{2n_{i-1}} \xrightarrow{F_1} \mathbf{4n_{i-1}}
$$

1.  Start with $n_{i-2} \equiv \mathbf{1 \pmod{4}}$ (odd).
2.  $\mathbf{F_1(n_{i-2})} = 2n_{i-2} \equiv \mathbf{2 \pmod{8}}$.
3.  $\mathbf{F_1(2n_{i-2})} = 4n_{i-2} \equiv \mathbf{4 \pmod{8}}$. (This is a $4 \times \text{Odd}$ number).

Since the inverse Collatz graph guarantees that any number $n$ has infinite parents of the form $2n$, the $\mathbf{F_1}$ operation perpetually generates $4 \pmod{8}$ numbers from $2 \pmod{8}$ numbers from $1 \pmod{4}$ numbers. Since the $1 \pmod{4}$ numbers are generated in every generation, their image under $\mathbf{F_1 \circ F_1}$ ensures the continuous presence of $4 \pmod{8}$ numbers.

#### B. Generation of $5 \pmod{8}$ Numbers

A number $n'$ is $5 \pmod{8}$ if $n' = 8k + 5$. This is an odd number.

Odd numbers are only generated by the $\mathbf{F_0}$ operation, as $\mathbf{F_1}$ only generates even numbers.
$$F_0(n) = \frac{2n-1}{3}$$

The rule for $F_0$ to result in an integer is $\mathbf{n \equiv 2 \pmod{3}}$.

We check the parity of $F_0(n)$ based on $n \pmod{8}$:
$$F_0(n) \pmod{8} \equiv \frac{2n-1}{3} \pmod{8}$$

| Input $n$ | $n \pmod{3}$ | $n \pmod{8}$ | $2n-1 \pmod{24}$ | $\frac{2n-1}{3} \pmod{8}$ |
| :---: | :---: | :---: | :---: | :---: |
| $n = 3k+2$ | $2 \pmod{3}$ | $2 \pmod{8}$ | $2(2)-1 = 3$ | $\mathbf{1 \pmod{8}}$ |
| $n = 3k+2$ | $2 \pmod{3}$ | $5 \pmod{8}$ | $2(5)-1 = 9$ | $\mathbf{3 \pmod{8}}$ |
| $n = 3k+2$ | $2 \pmod{3}$ | $8 \pmod{8} \equiv 0 \pmod{8}$ | $2(8)-1 = 15$ | $\mathbf{5 \pmod{8}}$ |
| $n = 3k+2$ | $2 \pmod{3}$ | $11 \pmod{8} \equiv 3 \pmod{8}$ | $2(11)-1 = 21$ | $\mathbf{7 \pmod{8}}$ |

The **structural feature** that causes the generation of $5 \pmod{8}$ numbers in every generation is the **periodic nature of the $F_0$ operation's output parity (mod 8).**

Specifically:
* The $F_0$ operation requires $\mathbf{n \equiv 0 \pmod{8}}$ as input to produce $\mathbf{5 \pmod{8}}$ output.
* $n \equiv 0 \pmod{8}$ numbers are generated by $\mathbf{F_1 \circ F_1 \circ F_1}$ (i.e., $8n_{i-3}$).
* Since the lattice continuously generates new $1 \pmod{4}$ numbers, it continuously generates $0 \pmod{8}$ numbers as their descendants.
* Crucially, every integer $n'$ has an $\mathbf{F_0}$ parent if $\mathbf{n' \equiv 5 \pmod{8}}$ (since $8k+5 \equiv 2 \pmod{3}$). This ensures the set is never empty.

### 6. Summary of the Structural Feature

The systematic generation of $4 \pmod{8}$ and $5 \pmod{8}$ numbers in every generation is caused by the **Interplay between the $\mathbf{F_1}$ function and the modular constraints of the $\mathbf{F_0}$ function:**

| Modulo Class | Structural Cause | MR Path Structure |
| :---: | :---: | :---: |
| $\mathbf{4 \pmod{8}}$ | **Repeated $\mathbf{F_1}$ application ($\mathbf{2n}$)** on $1 \pmod{4}$ numbers. | $\dots \xrightarrow{F_1} 2 \pmod{8} \xrightarrow{F_1} \mathbf{4 \pmod{8}}$ |
| $\mathbf{5 \pmod{8}}$ | **The $\mathbf{F_0}$ function's $\mathbf{n \equiv 0 \pmod{8}}$ input requirement.** | $\dots \xrightarrow{F_1} 0 \pmod{8} \xrightarrow{F_0} \mathbf{5 \pmod{8}}$ |

The fact that $1 \pmod{4}$ numbers are always present guarantees a continuous supply of $4 \pmod{8}$ and $5 \pmod{8}$ numbers via short, deterministic sequences of the $F_0$ and $F_1$ generators.

Would you like to explore which generation function sequence ($\mathbf{F_0}$ and $\mathbf{F_1}$) corresponds to the **non-neighborly** $7 \pmod{8}$ numbers to understand why their labels become so long?

In [202]:
for j in range(5, 50, 8):
    i = j - 1
    label_4 = collatzPath(i)
    label_5 = collatzPath(j)
    print(label_4)
    print(label_5)
    print("-----")

11
0111
-----
1100111
0110111
-----
110111
011111
-----
1100010110111
0110010110111
-----
110100010110111
011100010110111
-----
110010110111
011010110111
-----


In [211]:
# In general when we swap '011' for '110' WE ALWAYS GET 1 even from Rationals
for label_tail in generationLabels(8):
    label_4 = "110" + label_tail
    label_5 = "011" + label_tail
    val_4 = mrTupValue(mrTupFromPath(label_4))
    val_5 = mrTupValue(mrTupFromPath(label_5))
    val_diff = Fraction(*val_5) - Fraction(*val_4)
    if val_diff != 1:
        print((label_4, label_5))

$$
F_{tail}(F_0(F_1(F_1(N_{\emptyset})))) -1 = F_{tail}(F_1(F_1(F_0(N_{\emptyset})))) 
$$

$$
F_0(F_1(F_1(N_{\emptyset}))) -1 = F_1(F_1(F_0(N_{\emptyset})))
$$

$$
V(T_{011}) -1 = V(T_{110})
$$

$$ \frac{7}{3} - 1 = \frac{4}{3} $$


# $3n+2$ Neighborly label transform

$n' = 3n+2$ Can be accomplished WHEN the label starts with $101$ by changing the start to $111$

In [220]:
for label_tail in generationLabels(8):
    label_101 = "101" + label_tail + "111"
    label_111 = "111" + label_tail + "111"    
    val_101 = mrTupValue(mrTupFromPath(label_101))
    if val_101[1] == 1:
        val_111 = mrTupValue(mrTupFromPath(label_111))
        print((len(label_101), val_101[0], val_111[0], val_101[0] % 8, val_111[0] % 8, label_101, label_111))
#
for label_tail in generationLabels(9):
    label_101 = "101" + label_tail + "111"
    label_111 = "111" + label_tail + "111"    
    val_101 = mrTupValue(mrTupFromPath(label_101))
    if val_101[1] == 1:
        val_111 = mrTupValue(mrTupFromPath(label_111))
        print((len(label_101), val_101[0], val_111[0], val_101[0] % 8, val_111[0] % 8, label_101, label_111))



(14, 1706, 5120, 2, 0, '10111111110111', '11111111110111')
(14, 554, 1664, 2, 0, '10111110110111', '11111110110111')
(14, 1818, 5456, 2, 0, '10110111111111', '11110111111111')
(14, 602, 1808, 2, 0, '10110101111111', '11110101111111')
(14, 186, 560, 2, 0, '10110011110111', '11110011110111')
(14, 58, 176, 2, 0, '10110010110111', '11110010110111')
(14, 18, 56, 2, 0, '10100010110111', '11100010110111')
(15, 10922, 32768, 2, 0, '101111111111111', '111111111111111')
(15, 3626, 10880, 2, 0, '101111101111111', '111111101111111')
(15, 1130, 3392, 2, 0, '101111011110111', '111111011110111')
(15, 362, 1088, 2, 0, '101111010110111', '111111010110111')
(15, 1210, 3632, 2, 0, '101100111111111', '111100111111111')
(15, 122, 368, 2, 0, '101100011110111', '111100011110111')
(15, 402, 1208, 2, 0, '101000111111111', '111000111111111')


In [226]:
# Do 2(mod 8) numbers always start with 101?  YES
for i in range(10, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "101":
        print((i, label))

In [227]:
# Do 0(mod 8) numbers always start with 111?  YES
for i in range(16, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "111":
        print((i, label))

In [229]:
# Do 1(mod 8) numbers always start with 010?  YES
for i in range(17, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "010":
        print((i, label))

In [232]:
# Do 7(mod 8) numbers always start with 000?  Yes
for i in range(15, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "000":
        print((i, label))

In [234]:
# Do 4(mod 8) numbers always start with 110?  Yes
for i in range(12, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "110":
        print((i, label))

In [236]:
# Do 6(mod 8) numbers always start with 100?  Yes
for i in range(14, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "100":
        print((i, label))

In [239]:
# Do 3(mod 8) numbers always start with 001?  Yes
for i in range(11, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "001":
        print((i, label))

In [242]:
# Do 5(mod 8) numbers always start with 011?  Yes
for i in range(13, 2000, 8):
    label = collatzPath(i)
    if label[0:3] != "011":
        print((i, label))

In [245]:
# Since 7 (mod 8) is the slowest grower, can we always convert it to everything else? YES!!!
PREFIXES =  ["000", "001", "010", "011", "100", "101", "110", "111"]
status =[[0]*8, [0]*8]
for i in range(15, 10000, 8):
    label = collatzPath(i)
    for prefix in [1,2,3,4,5,6,7]:
        label_ = PREFIXES[prefix] + label
        val_ = mrTupValue(mrTupFromPath(label))
        if val_[1] == 1:
            status[1][prefix] += 1
        else:
            status[0][prefix] += 1
status


[[0, 0, 0, 0, 0, 0, 0, 0], [0, 1249, 1249, 1249, 1249, 1249, 1249, 1249]]

In [249]:
# Since 7 (mod 8) is the slowest grower, can we always convert it to everything else? YES!!!
PREFIXES =  ["111", "010", "101", "001", "110", "011", "100", "000"]
MOD =       [  0,     1,     2,     3,     4,     5,     6,     7  ]
for j in range(8):
    k = MOD[j]
    status =[[0]*8, [0]*8]
    for i in range(k+16, 10000, 8):
        label = collatzPath(i)
        for prefix in range(8):
            label_ = PREFIXES[prefix] + label[3:]
            val_ = mrTupValue(mrTupFromPath(label_))
            if val_[1] == 1:
                status[1][prefix] += 1
            else:
                status[0][prefix] += 1
    print((k, PREFIXES[j], status))


(0, '111', [[0, 1109, 832, 1109, 832, 832, 1110, 1202], [1248, 139, 416, 139, 416, 416, 138, 46]])
(1, '010', [[0, 0, 1248, 1248, 0, 0, 1248, 1248], [1248, 1248, 0, 0, 1248, 1248, 0, 0]])
(2, '101', [[0, 1248, 0, 832, 1248, 1248, 1248, 1248], [1248, 0, 1248, 416, 0, 0, 0, 0]])
(3, '001', [[0, 1248, 0, 0, 1248, 1248, 1248, 1248], [1248, 0, 1248, 1248, 0, 0, 0, 0]])
(4, '110', [[0, 832, 1248, 1248, 0, 0, 832, 1110], [1248, 416, 0, 0, 1248, 1248, 416, 138]])
(5, '011', [[0, 832, 1248, 1248, 0, 0, 832, 1110], [1248, 416, 0, 0, 1248, 1248, 416, 138]])
(6, '100', [[0, 1248, 1248, 1248, 0, 0, 0, 832], [1248, 0, 0, 0, 1248, 1248, 1248, 416]])
(7, '000', [[0, 1248, 1248, 1248, 0, 0, 0, 0], [1248, 0, 0, 0, 1248, 1248, 1248, 1248]])


# 1:1 mappings through 3 bit prefix changes:
- 0 (mod 8):
- 1 (mod 8): 0, 4, 5
- 2 (mod 8): 0
- 3 (mod 8): 0, 2
- 4 (mod 8): 0, 4, 5, 6
- 5 (mod 8): 0, 6
- 6 (mod 8): 0, 4, 5
- 7 (mod 8): 0, 4, 5, 6
- 

In [250]:
# Since 7 (mod 8) is the slowest grower, can we always convert it to everything else? YES!!!
PREFIXES =  ["111", "010", "101", "001", "110", "011", "100", "000"]
MOD =       [  0,     1,     2,     3,     4,     5,     6,     7  ]
for j in range(8):
    k = MOD[j]
    status =[[0]*8, [0]*8]
    for i in range(k+16, 10000, 8):
        label = collatzPath(i)
        for prefix in range(8):
            label_ = PREFIXES[prefix] + "1" + label[3:]
            val_ = mrTupValue(mrTupFromPath(label_))
            if val_[1] == 1:
                status[1][prefix] += 1
            else:
                status[0][prefix] += 1
    print((k, PREFIXES[j], status))


(0, '111', [[0, 1110, 832, 1109, 832, 832, 1109, 1202], [1248, 138, 416, 139, 416, 416, 139, 46]])
(1, '010', [[0, 1248, 0, 0, 1248, 1248, 1248, 1248], [1248, 0, 1248, 1248, 0, 0, 0, 0]])
(2, '101', [[0, 832, 1248, 1248, 0, 0, 832, 1109], [1248, 416, 0, 0, 1248, 1248, 416, 139]])
(3, '001', [[0, 1248, 1248, 1248, 0, 0, 0, 832], [1248, 0, 0, 0, 1248, 1248, 1248, 416]])
(4, '110', [[0, 1248, 0, 832, 1248, 1248, 1248, 1248], [1248, 0, 1248, 416, 0, 0, 0, 0]])
(5, '011', [[0, 1248, 0, 832, 1248, 1248, 1248, 1248], [1248, 0, 1248, 416, 0, 0, 0, 0]])
(6, '100', [[0, 1248, 0, 1248, 1248, 1248, 1248, 1248], [1248, 0, 1248, 0, 0, 0, 0, 0]])
(7, '000', [[0, 1248, 0, 1248, 1248, 1248, 1248, 1248], [1248, 0, 1248, 0, 0, 0, 0, 0]])


# 1:1 mappings through (3 bit + "1") prefix changes:
- 0 (mod 8):
- 1 (mod 8): 2, 3
- 2 (mod 8): 0, 4, 5
- 3 (mod 8): 0, 4, 5, 6
- 4 (mod 8): 0, 2
- 5 (mod 8): 0, 2
- 6 (mod 8): 0, 2
- 7 (mod 8): 0, 2
- 

In [251]:
# Since 7 (mod 8) is the slowest grower, can we always convert it to everything else? YES!!!
PREFIXES =  ["111", "010", "101", "001", "110", "011", "100", "000"]
MOD =       [  0,     1,     2,     3,     4,     5,     6,     7  ]
for j in range(8):
    k = MOD[j]
    status =[[0]*8, [0]*8]
    for i in range(k+16, 10000, 8):
        label = collatzPath(i)
        for prefix in range(8):
            label_ = "1" + PREFIXES[prefix] + label[3:]
            val_ = mrTupValue(mrTupFromPath(label_))
            if val_[1] == 1:
                status[1][prefix] += 1
            else:
                status[0][prefix] += 1
    print((k, PREFIXES[j], status))


(0, '111', [[0, 1109, 832, 1109, 832, 832, 1110, 1202], [1248, 139, 416, 139, 416, 416, 138, 46]])
(1, '010', [[0, 0, 1248, 1248, 0, 0, 1248, 1248], [1248, 1248, 0, 0, 1248, 1248, 0, 0]])
(2, '101', [[0, 1248, 0, 832, 1248, 1248, 1248, 1248], [1248, 0, 1248, 416, 0, 0, 0, 0]])
(3, '001', [[0, 1248, 0, 0, 1248, 1248, 1248, 1248], [1248, 0, 1248, 1248, 0, 0, 0, 0]])
(4, '110', [[0, 832, 1248, 1248, 0, 0, 832, 1110], [1248, 416, 0, 0, 1248, 1248, 416, 138]])
(5, '011', [[0, 832, 1248, 1248, 0, 0, 832, 1110], [1248, 416, 0, 0, 1248, 1248, 416, 138]])
(6, '100', [[0, 1248, 1248, 1248, 0, 0, 0, 832], [1248, 0, 0, 0, 1248, 1248, 1248, 416]])
(7, '000', [[0, 1248, 1248, 1248, 0, 0, 0, 0], [1248, 0, 0, 0, 1248, 1248, 1248, 1248]])


# 1:1 mappings through ("1" + 3 bit) prefix changes:
- 0 (mod 8):
- 1 (mod 8): 0, 4, 5
- 2 (mod 8): 0, 2
- 3 (mod 8): 0, 4, 5, 6
- 4 (mod 8): 0, 5, 6
- 5 (mod 8): 0, 4
- 6 (mod 8): 0, 4, 5
- 7 (mod 8): 0, 4, 5, 6


In [223]:
for label_tail in generationLabels(8):
    label_101 = "101" + label_tail + "111"
    label_111 = "0111" + label_tail + "111"    
    val_101 = mrTupValue(mrTupFromPath(label_101))
    if val_101[1] == 1:
        val_111 = mrTupValue(mrTupFromPath(label_111))
        print((len(label_101), val_101[0], val_111[0], val_101[0] % 8, val_111[0] % 8, label_101, label_111))
#
for label_tail in generationLabels(9):
    label_101 = "101" + label_tail + "111"
    label_111 = "0111" + label_tail + "111"    
    val_101 = mrTupValue(mrTupFromPath(label_101))
    if val_101[1] == 1:
        val_111 = mrTupValue(mrTupFromPath(label_111))
        print((len(label_101), val_101[0], val_111[0], val_101[0] % 8, val_111[0] % 8, label_101, label_111))


(14, 1706, 3413, 2, 5, '10111111110111', '011111111110111')
(14, 554, 1109, 2, 5, '10111110110111', '011111110110111')
(14, 1818, 3637, 2, 5, '10110111111111', '011110111111111')
(14, 602, 1205, 2, 5, '10110101111111', '011110101111111')
(14, 186, 373, 2, 5, '10110011110111', '011110011110111')
(14, 58, 117, 2, 5, '10110010110111', '011110010110111')
(14, 18, 37, 2, 5, '10100010110111', '011100010110111')
(15, 10922, 21845, 2, 5, '101111111111111', '0111111111111111')
(15, 3626, 7253, 2, 5, '101111101111111', '0111111101111111')
(15, 1130, 2261, 2, 5, '101111011110111', '0111111011110111')
(15, 362, 725, 2, 5, '101111010110111', '0111111010110111')
(15, 1210, 2421, 2, 5, '101100111111111', '0111100111111111')
(15, 122, 245, 2, 5, '101100011110111', '0111100011110111')
(15, 402, 805, 2, 5, '101000111111111', '0111000111111111')


# The values of the 3bit mrTups

In [263]:
def changePrefix(label, n, new_prefix):
    return new_prefix + label[n:]
#

In [261]:
for n_mod_8 in MOD:
    prefix = PREFIXES[n_mod_8]
    val = mrTupValue(mrTupFromPath(prefix))
    print(f"{n_mod_8}\t{prefix}\t{val}")

0	111	(8, 1)
1	010	(1, 9)
2	101	(2, 1)
3	001	(1, 3)
4	110	(4, 3)
5	011	(7, 3)
6	100	(-2, 9)
7	000	(-11, 27)


# Formulas for all 1:1 3 bit substitutions

## 1 (mod 8): 0, 4, 5
- 0:  1/9 -> 8   ; (17 -> 160, 25 -> 232) ; $9 \cdot T_{010} + 7 \mapsto T_{111}$
- 4:  1/9 -> 4/3 ; (17 -> 52,  25 -> 76) ; $3 \cdot T_{010} + 1 \mapsto T_{110}$
- 5:  1/9 -> 7/3 ; (17 -> 53,  25 -> 77) ; $3 \cdot T_{010} + 2 \mapsto T_{011}$
## 2 (mod 8): 0
- 0:  2 -> 8   ; (18 -> 56, 26 -> 80) ; $3 \cdot T_{101} +2 -> T_{111}$
## 3 (mod 8): 0, 2
- 0:  1/3 -> 8   ; (19 -> 176, 27 -> 248) ; 9*T_001 + 5 -> T_111
- 2:  1/3 -> 2   ; (19 -> 58, 27 -> 82) ; 3*T_001 + 1 -> T_101
## 4 (mod 8): 0, 5
- 0:  4/3 -> 8   ; (20 -> 64, 28 -> 88) ; 3*T_110 + 4 -> T_111
- 5:  4/3 -> 7/3 ; (20 -> 21 , 28 -> 29) ; T_110 + 1 -> T_011
- 6 : ??
## 5 (mod 8): 0
- 0:  7/3 -> 8     ; (21 -> 64, 29 -> 88) ; 3*T_011 + 1 -> T_111
- 6 : ??
## 6 (mod 8): 0, 4, 5
- 0:  -2/9 -> 8   ; (22 -> 208, 30 -> 280) ; $9 \cdot T_{100} + 10 \mapsto T_{111}$
- 4:  -2/9 -> 4/3 ; (22 -> 68, 30 -> 92) ; $3 \cdot T_{100} + 2 \mapsto T_{110}$
- 5:  -2/9 -> 7/3 ; (22 -> 69, 30 -> 03) ; $3 \cdot T_{100} + 3 \mapsto T_{011}$
## 7 (mod 8): 0, 4, 5, 6
- 0:  -11/27 -> 8   ; (23 -> 640, 30 -> 856) ; $27 \cdot T_{000} + 19 \mapsto T_{111}$
- 4:  -11/27 -> 4/3 ; (23 -> 212, 30 -> 284) ; $9 \cdot T_{000} + 5 \mapsto T_{110}$
- 5:  -11/27 -> 7/3 ; (23 -> 213, 30 -> 285) ; $9 \cdot T_{000} + 6 \mapsto T_{011}$
- 6:  -11/27 ->-2/9 ; (23 -> , 30 -> ) ; $3 \cdot T_{000} + 3 \mapsto T_{100}$

In [321]:
20 %24, 44 % 24

(20, 20)

In [362]:
def T_010_to_T_111(k):
    return 9*k + 7
#
def T_010_to_T_110(k):
    return 3*k + 1
#
def T_010_to_T_011(k):
    return 3*k + 2
#
def T_101_to_T_111(k):
    return 3*k + 2
#
def T_001_to_T_111(k):
    return 9*k + 5
#
def T_001_to_T_101(k):
    return 3*k + 1
#
def T_110_to_T_111(k):
    return 3*k + 4
#
def T_110_to_T_011(k):
    return k + 1
#
def T_110_to_T_100(k):
    # TODO: Not correct
    if (k % 24) == 20:
        return (k-2)//3
    else:
        return k-2
#
def T_011_to_T_111(k):
    return 3*k + 1
#
def T_011_to_T_100(k):
    if (k%24) == 21:
        return (k//3) -1
    else:
        return k+2
#
def T_100_to_T_111(k):
    return 9*k + 10
#
def T_100_to_T_110(k):
    return 3*k + 2
#
def T_100_to_T_011(k):
    return 3*k + 3
#
def T_000_to_T_111(k):
    return 27*k + 19
#
def T_000_to_T_110(k):
    return 9*k + 5
#
def T_000_to_T_011(k):
    return 9*k + 6
#
def T_000_to_T_100(k):
    return 3*k + 1
#


def checkFormula(mod_8, tranform_func, prefix):
    for i in range(1000):
        k = 8*(i+2) + mod_8
        label = collatzPath(k)
        label_ = prefix + label[len(prefix):]
        val_ = mrTupValue(mrTupFromPath(label_))
        if val_[1] != 1:
            print(f"Label substitution FAILED {k}({label}) -> ({label_}) has noninteger {val_}")
            return False
        k_ = tranform_func(k) 
        if k_ != val_[0]:
            print(f"{tranform_func.__name__} FAILED {k} should -> {val_} but function gave {k_}")
            return False
    return True
#
       

In [363]:
checkFormula(1, T_010_to_T_111, "111")
checkFormula(1, T_010_to_T_110, "110")
checkFormula(1, T_010_to_T_011, "011")

checkFormula(2, T_101_to_T_111, "111")
checkFormula(3, T_001_to_T_111, "111")

checkFormula(3, T_001_to_T_101, "101")

checkFormula(4, T_110_to_T_111, "111")
checkFormula(4, T_110_to_T_011, "011")
#checkFormula(4, T_110_to_T_100, "100")  removed 4(mod 8) -> 6(mod 8) case

checkFormula(5, T_011_to_T_111, "111")
# checkFormula(5, T_011_to_T_100, "100") removed 5(mod 8) -> 6(mod 8) case

checkFormula(6, T_100_to_T_111, "111")
checkFormula(6, T_100_to_T_110, "110")
checkFormula(6, T_100_to_T_011, "011")

checkFormula(7, T_000_to_T_111, "111")
checkFormula(7, T_000_to_T_110, "110")
checkFormula(7, T_000_to_T_011, "011")
checkFormula(7, T_000_to_T_100, "100")

True

True

In [262]:
collatzPath(17)

'010110111'

In [361]:
prefix = "100"
mod_8 = 7
mrTupValue(mrTupFromPath(changePrefix(collatzPath(16 + mod_8), 3, prefix))), \
    mrTupValue(mrTupFromPath(changePrefix(collatzPath(24 + mod_8), 3, prefix))), \
    mrTupValue(mrTupFromPath(changePrefix(collatzPath(32 + mod_8), 3, prefix))), \
    mrTupValue(mrTupFromPath(changePrefix(collatzPath(40 + mod_8), 3, prefix)))


((70, 1), (94, 1), (118, 1), (142, 1))

# Integer Generation Rate per Lattice Segment

We know that in the bottom of the Lattice it is harder to generate integers.  

The hypothesis is that each segment of the lattice at generation a' will generate integers at a rate similar to another segment in generation a.  In other words, the slower generators just lag.


In [255]:
def gen_rates_func():
    gen_rates = {}
    for a in range(4, 27, 1):
        gen_rates[a] = {}
        for p in PREFIXES:
            gen_rates[a][p] = 0
        for label in generationLabels(a):
            val = mrTupValue(mrTupFromPath(label))
            if val[1] == 1:
                gen_rates[a][label[0:3]] += 1
    return gen_rates    
#

# This is slow, pastes below
#gen_rates = gen_rates_func()

gen_rates_comment="""
{4: {'111': 1,
  '010': 1,
  '101': 0,
  '001': 0,
  '110': 1,
  '011': 1,
  '100': 0,
  '000': 0},
 5: {'111': 2,
  '010': 0,
  '101': 2,
  '001': 1,
  '110': 0,
  '011': 0,
  '100': 0,
  '000': 0},
 6: {'111': 2,
  '010': 1,
  '101': 0,
  '001': 0,
  '110': 2,
  '011': 2,
  '100': 1,
  '000': 0},
 7: {'111': 4,
  '010': 0,
  '101': 3,
  '001': 1,
  '110': 1,
  '011': 1,
  '100': 0,
  '000': 0},
 8: {'111': 5,
  '010': 1,
  '101': 1,
  '001': 0,
  '110': 3,
  '011': 3,
  '100': 1,
  '000': 0},
 9: {'111': 8,
  '010': 1,
  '101': 4,
  '001': 1,
  '110': 2,
  '011': 2,
  '100': 0,
  '000': 0},
 10: {'111': 10,
  '010': 2,
  '101': 3,
  '001': 2,
  '110': 4,
  '011': 4,
  '100': 1,
  '000': 0},
 11: {'111': 14,
  '010': 1,
  '101': 6,
  '001': 3,
  '110': 4,
  '011': 4,
  '100': 2,
  '000': 2},
 12: {'111': 18,
  '010': 2,
  '101': 5,
  '001': 2,
  '110': 8,
  '011': 8,
  '100': 5,
  '000': 2},
 13: {'111': 26,
  '010': 2,
  '101': 10,
  '001': 3,
  '110': 10,
  '011': 10,
  '100': 4,
  '000': 2},
 14: {'111': 36,
  '010': 3,
  '101': 12,
  '001': 3,
  '110': 14,
  '011': 14,
  '100': 5,
  '000': 2},
 15: {'111': 50,
  '010': 5,
  '101': 17,
  '001': 4,
  '110': 17,
  '011': 17,
  '100': 5,
  '000': 2},
 16: {'111': 67,
  '010': 9,
  '101': 22,
  '001': 7,
  '110': 22,
  '011': 22,
  '100': 6,
  '000': 2},
 17: {'111': 89,
  '010': 9,
  '101': 31,
  '001': 11,
  '110': 28,
  '011': 28,
  '100': 9,
  '000': 3},
 18: {'111': 117,
  '010': 14,
  '101': 37,
  '001': 12,
  '110': 40,
  '011': 40,
  '100': 14,
  '000': 3},
 19: {'111': 157,
  '010': 18,
  '101': 54,
  '001': 17,
  '110': 51,
  '011': 51,
  '100': 15,
  '000': 4},
 20: {'111': 208,
  '010': 23,
  '101': 69,
  '001': 25,
  '110': 69,
  '011': 69,
  '100': 21,
  '000': 4},
 21: {'111': 277,
  '010': 34,
  '101': 92,
  '001': 28,
  '110': 90,
  '011': 90,
  '100': 29,
  '000': 9},
 22: {'111': 367,
  '010': 44,
  '101': 124,
  '001': 47,
  '110': 121,
  '011': 121,
  '100': 37,
  '000': 8},
 23: {'111': 488,
  '010': 53,
  '101': 165,
  '001': 53,
  '110': 161,
  '011': 161,
  '100': 55,
  '000': 18},
 24: {'111': 649,
  '010': 74,
  '101': 214,
  '001': 71,
  '110': 220,
  '011': 220,
  '100': 71,
  '000': 15},
 25: {'111': 869,
  '010': 100,
  '101': 294,
  '001': 93,
  '110': 285,
  '011': 285,
  '100': 86,
  '000': 27},
 26: {'111': 1154,
  '010': 135,
  '101': 385,
  '001': 134,
  '110': 380,
  '011': 380,
  '100': 120,
  '000': 33}}
  """

In [258]:
for p in PREFIXES:
    production = []
    for a in gen_rates:
        production.append(str(gen_rates[a][p]))
    print("\t".join(production))

1	2	2	4	5	8	10	14	18	26	36	50	67	89	117	157	208	277	367	488	649	869	1154
1	0	1	0	1	1	2	1	2	2	3	5	9	9	14	18	23	34	44	53	74	100	135
0	2	0	3	1	4	3	6	5	10	12	17	22	31	37	54	69	92	124	165	214	294	385
0	1	0	1	0	1	2	3	2	3	3	4	7	11	12	17	25	28	47	53	71	93	134
1	0	2	1	3	2	4	4	8	10	14	17	22	28	40	51	69	90	121	161	220	285	380
1	0	2	1	3	2	4	4	8	10	14	17	22	28	40	51	69	90	121	161	220	285	380
0	0	1	0	1	0	1	2	5	4	5	5	6	9	14	15	21	29	37	55	71	86	120
0	0	0	0	0	0	0	2	2	2	2	2	2	3	3	4	4	9	8	18	15	27	33


# What does 0(mod 8) map to when we simply remove leading ones?

If we have a number 111..., then what other numbers 11..., 1..., ... do we HAVE to HAVE?


In [260]:
print("\t".join(PREFIXES))

111	010	101	001	110	011	100	000


In [224]:
1706 * 2 - 3413

-1

In [214]:
mrTupValue(mrTupFromPath("011"))

(7, 3)

In [215]:
mrTupValue(mrTupFromPath("110"))

(4, 3)

In [205]:
for i in range(12, 2000, 8):
    label_4 = collatzPath(i)
    if label_4[0:3] != "110":
        print(i, label_4)

In [164]:
for a in range(4,16,1):
    """ Count the number of 7 (mod 8) numbers found per generation.
    """
    accum = [0]*8
    int_count = 0
    for label in generationLabels(a):
        T = mrTupFromPath(label)
        val = mrTupValue(T)
        if val[1] == 1:
            int_count += 1
            accum[(val[0] % 8)] += 1
            if (val[0] % 8) == 4:
                print((val[0] % 8, a, val, label, T))
            if (val[0] % 8) == 5:
                print((val[0] % 8, a, val, label, T))
    print((a, accum, int_count))
#


(4, 4, (4, 1), '1101', ((4, -1), [(2, -1)]))
(5, 4, (5, 1), '0111', ((4, -1), [(0, -1)]))
(4, [1, 1, 0, 0, 1, 1, 0, 0], 4)
(5, [2, 0, 2, 1, 0, 0, 0, 0], 5)
(4, 6, (20, 1), '110111', ((6, -1), [(2, -1)]))
(4, 6, (4, 1), '110101', ((6, -2), [(2, -1), (4, -2)]))
(5, 6, (21, 1), '011111', ((6, -1), [(0, -1)]))
(5, 6, (5, 1), '011101', ((6, -2), [(0, -1), (4, -2)]))
(6, [2, 1, 0, 0, 2, 2, 1, 0], 8)
(4, 7, (12, 1), '1100111', ((7, -2), [(2, -1), (3, -2)]))
(5, 7, (13, 1), '0110111', ((7, -2), [(0, -1), (3, -2)]))
(7, [4, 0, 3, 1, 1, 1, 0, 0], 10)
(4, 8, (84, 1), '11011111', ((8, -1), [(2, -1)]))
(4, 8, (20, 1), '11011101', ((8, -2), [(2, -1), (6, -2)]))
(4, 8, (4, 1), '11010101', ((8, -3), [(2, -1), (4, -2), (6, -3)]))
(5, 8, (85, 1), '01111111', ((8, -1), [(0, -1)]))
(5, 8, (21, 1), '01111101', ((8, -2), [(0, -1), (6, -2)]))
(5, 8, (5, 1), '01110101', ((8, -3), [(0, -1), (4, -2), (6, -3)]))
(8, [5, 1, 1, 0, 3, 3, 1, 0], 14)
(4, 9, (52, 1), '110110111', ((9, -2), [(2, -1), (5, -2)]))
(4, 9, 

In [186]:
val = mrTupValue(mrTupFromPath("0000010100100010000101100010010000001100001110101011101100011110"))
(val, val[0]/val[1])

((-3497998772378597035, 4052555153018976267), -0.8631588319711678)

In [194]:

val = mrTupValue(mrTupFromPath("011010"))
(val, val[0]/val[1])

((-1, 27), -0.037037037037037035)